---
title: "MCAO_Deconvolutions"
author: "Daniel Zucha"
date: "2023-06-04"
output: html_document
---

Markdown for the deconvolution analysis of the spatial transcriptomics dataset. Contains the preparation of the reference dataset from Milich et al. (2021) for the reference-based deconvolution (RCTD), as well as the reference-free deconvolution approach (STdeconvolve).

Essential input files:
  - Milich et al reference dataset: for processing from scratch, download the inputs from their GEO (GSE162610). To use the re-analyzed dataset by us, download "Milich2021_Batch2_seurat_sct_RedDim.Rds" from our Mendeley data repository. For the milich data, create a subfolder named 'Milich_2021' in the 'data' folder.
  - STdeconvolve: "STdeconvolve_ResultsList_chosenKs.Rds", "STdeconvolve_markers.zip" (the topic markers)
  - spatial: seurat_spatial_integrated_ready.Rds
  - metadata gene lists: gene lists from selected publications (available as Supp. table 4)


=== Essential chunks. Always load. ===

```{r libraries}
library(Seurat); library(tidyverse); library(ggplot2); library(tibble); library(openxlsx); library(RColorBrewer); library(magrittr); library(stringr); library(gridExtra); library(Matrix); library(patchwork); library(ggrepel)

# sourcing custom functions and color palettes
source("code/supporting_functions_MCAO.R")
```

```{r working space and lists}
# ws
if(!dir.exists("ws/deconvolution")){dir.create("ws/deconvolution")
  ws <- "ws/deconvolution"}else{ws <- "ws/deconvolution"}

milich.source <- "data/Metadata/Milich2021/"

# lists
plot.list <- list()
result.list <- list()
```

=== Ref-based deconvolution: RCTD using Zeng 2023 ref ===
Next, we will use the annotated reference dataset for a deconvolution of in-house generated spatial data.

```{r load libraries and datasets}
library(spacexr)

# single-cell reference
seurat <- readRDS(file = "data/Metadata/seurat_Zeng2023_qcfiltered_normalized_integrated_annotated.rds")

# spatial seurat
spatial.seurat <- readRDS(file = "data/seurat_spatial_integrated_ready.Rds")
DefaultAssay(spatial.seurat) <- "Spatial"
```

Single-cell reference object prep
```{r Reference}
# counts and metadata information
counts <- seurat[["RNA"]]$counts
meta_data <- data.frame("Barcodes" = Cells(seurat),
                        "Cluster" = seurat$CellTypeClass, # annotated celltypes
                        "nUMI" = colSums(counts))
cell_types <- meta_data$Cluster %>% as.factor(); names(cell_types) <- meta_data$Barcodes # assign barcode names to the individual reference cells
nUMI <- meta_data$nUMI; names(nUMI) <- meta_data$Barcodes # create named nUMI list

### Create the Reference object
reference <- Reference(counts, cell_types, nUMI, min_UMI = 300)

## (optional) Examine reference object
print(dim(reference@counts)) #observe Gene Expression matrix
table(reference@cell_types) #number of occurences for each cell type
```

Spatial object prep
```{r Spatial data prep}
spatial.seurat@meta.data %>% glimpse
spatial.seurat@meta.data$Sample %>% table
sections <- c("Ctrl", "1DPI", "3DPI", "7DPI")

spatial.seurat %>% SpatialPlot(
  group.by = "DetailedRegionAnnoShort", 
  cols = col.list$cols_mono_short, 
  images = sections,
  image.alpha = 0, 
  stroke = 0, 
  crop = F) & NoLegend()

## individual sections are provided separately
counts <- list()
coords <- list()
nUMI.spatial <- list()
for(i in sections){
  counts[[i]] <- spatial.seurat[["Spatial"]]@counts[, spatial.seurat$Condition %in% i] # RNA counts per replicate
  coords[[i]] <- data.frame("SpatialBarcodes" = Cells(spatial.seurat)[spatial.seurat$Condition %in% i], #coordinates
                            "xcoord" = spatial.seurat@images[[i]]@coordinates$imagecol, 
                            "ycoord" = spatial.seurat@images[[i]]@coordinates$imagerow)
  rownames(coords[[i]]) <- coords[[i]]$SpatialBarcodes; coords[[i]]$SpatialBarcodes <- NULL   # spot barcodes to rownames
  nUMI.spatial[[i]] <- colSums(counts[[i]])
}

# Spatial RNA objects
puck <- list()
for(i in sections){
  puck[[i]] <- SpatialRNA(coords[[i]], counts[[i]], nUMI.spatial[[i]])
}
```

(Optional) Moving to a background job by running '0F_04_MCAO_RCTD_deconvolution_background.R'
```{r optional moving to a background job}
save(reference, puck, sections, file = file.path("data", "1DP_04_MCAO_rctd_zeng2023_background.Rdata"))
```

Run RCTD deconvolution
```{r RCTD object and deconvo}
# prepare RCTD object
puck %>% glimpse # individual spatialRNA objects inside puck list
sections ## names for the replicates to be analyzed

# Separate RCTD for sections (easier downstream analysis)
rm(nUMI.spatial, nUMI, counts, coords)
gc()
myRCTD.list <- list()
for(replicate in sections){
  myRCTD.list[[replicate]] <- create.RCTD(spatialRNA = puck[[replicate]], 
                                          reference = reference, max_cores = 32)
  }

celltypes_levels <- reference@cell_types %>% levels
gc() ## create deconvo-ready files

for(i in sections){
  myRCTD.list[[i]] <- run.RCTD(myRCTD.list[[i]], 
                               doublet_mode = "full")}  ## deconvo itself; mode = "full" as we expect multiple cell types present in each spot.

# save the output
save(myRCTD.list, sections, celltypes_levels, file = "data/RCTDlist_scRefZeng2023_fullmode.Rdata")
saveRDS(myRCTD.list, file = "data/RCTDlist_scRefZeng2023_fullmode.Rds")
```

(optional) load rctd results
```{r load rctd results}
# load(file = "data/RCTDlist_scRefZeng2023_fullmode.Rdata")
```

Add RCTD results to the spatial object
```{r add RCTD to spatial}
if(myRCTD.list[[1]]@config$RCTDmode == "multi"){ # processing multi-mode results
  cat(crayon::yellow("Adding RCTD multi mode results.", "\n"))
  subweight.list <- list()
  for(i in names(myRCTD.list)){
    len <- myRCTD.list[[i]]@results %>% length
    subweight.list[[i]] <- lapply(1:len, function(ii){
      myRCTD.list[[i]]@results[[ii]][["sub_weights"]]
      })
    names(subweight.list[[i]]) <- myRCTD.list[[i]]@spatialRNA@coords %>% rownames()
  }; subweight.list <- purrr::flatten(subweight.list) # barcode-named list of doubles with deconvo proportions
  
  temp <- lapply(celltypes_levels, function(x){
    var <- which(sapply(subweight.list, function(y){x %in% names(y)})) # which spots include given cell type
    barcodes <- sapply(subweight.list[var], function(z){vals <- z[x]}) # barcode of the spot with cell type, extract celltype's value from that spot
    names(barcodes) <- names(barcodes) %>% stringr::str_remove(pattern = paste(".", x, sep = "")) # remove the '.celltypename' from barcode names
    notbarcodes <-  Cells(spatial.seurat)[!(Cells(spatial.seurat) %in% names(barcodes))] # barcodes for spots that do not contain given cell type
    zeros <- rep(0, length(notbarcodes)); names(zeros) <- notbarcodes # zero vector of notbarcodes length
    fullbarcodes <- c(barcodes, zeros) 
    return(fullbarcodes)  
  }); names(temp) <- celltypes_levels
  
  for(name in names(temp)){ # add the proportions to the spatial.seurat file
    spatial.seurat %<>% AddMetaData(temp[[name]], col.name = paste0("RCTD_Celltypes_", name)) # add deconvoluted values as metadata to the seurat object
  }
  
  rm(name, i, len)
  
} else if(myRCTD.list[[1]]@config$RCTDmode == "full"){
  cat(crayon::yellow("Adding RCTD full mode results.", "\n"))
  result.list[["full_weightlist"]] <- lapply(myRCTD.list, function(i){
    norm_weights <- i@results$weights %>% normalize_weights() %>% as.data.frame()
    return(norm_weights)
  }) %>% bind_rows()
  
  # reorder the weightlist by spatial seurat barcodes
  result.list[["full_weightlist"]] %<>% slice(match(rownames(result.list[["full_weightlist"]]), spatial.seurat %>% Cells))
  rctd_colnames <- paste0("RCTD_", colnames(result.list[["full_weightlist"]]))
  spatial.seurat %<>% AddMetaData(
    metadata = result.list[["full_weightlist"]], 
    col.name = rctd_colnames
    )
  
  # created missing values replace by zero
  spatial.seurat@meta.data[rctd_colnames][is.na(spatial.seurat@meta.data[rctd_colnames])] <- 0
} else {
  cat(crayon::yellow("Adding RCTD double mode results is not supported.", "\nCodebase is missing.", "\n"))
}
```

save spatial with RCTD data
```{r save spatial after deconvo}
saveRDS(spatial.seurat, file = file.path("data", "seurat_spatial_integrated_ready_rctdZeng2023.rds"))
```

```{r RCTD SpatialPlot}
# Spatial plots, rocket color gradient for all
temp <- lapply(
  paste0(celltypes_levels), ## RCTD results
  function(celltype_level){
    celltype <- paste0("RCTD_", celltype_level)
    plot <- spatial.seurat %>% SpatialFeaturePlot(
      features = celltype, 
      images = sections, 
      crop = F, 
      ncol = 4, 
      pt.size.factor = 1.5,
      stroke = 0, 
      alpha = 1, 
      image.alpha = 0, 
      combine = T
    ) & 
      theme(legend.position = "none") &
      viridis::scale_fill_viridis(
        option = 'rocket', 
        direction = -1, 
        limits = c(0,1)) & 
      labs(title = NULL)
    
    plot <- plot + plot_annotation(title = NULL, subtitle = celltype_level)
  }
)
names(temp) <- paste0("SpatialPlot_", celltypes_levels)
plot.list %<>% append(values = temp)

print(plot.list$SpatialPlot_VSMCs)
```

```{r Subset for cortex-only spatial.seurat}
## create a subset spatial seurat for roi
isch.spots <- spatial.seurat@meta.data %>% filter(DetailedRegionAnnoShort %in% roi) %>% rownames
ctx.spatial.seurat <- subset(spatial.seurat, cells = isch.spots)

ctx.spatial.seurat %>% SpatialPlot(
  features = "RCTD_Astrocytes", 
  images = sections, 
  alpha = 1, 
  image.alpha = 0.1, 
  crop = F, 
  stroke = 0, 
  combine = T, 
  ncol = 4
  ) & 
  theme(legend.position = "none") &
  viridis::scale_fill_viridis(
    option = 'rocket', 
    direction = -1, 
    limits = c(0,1)) & 
  labs(title = NULL)
```

```{r RCTD Stacked barplot}
# celltypes of interest
coi <- spatial.seurat@meta.data %>% colnames %>% stringr::str_subset(pattern = "^RCTD_")

# regions of interest - affected by stroke
Idents(spatial.seurat) <- "DetailedRegionAnnoShort"
roi <- c("CTX1-4", "CTX5", "CTX6", "ISD1c", "ISD1p", "ISD3c", "ISD3p", "ISD7c", "ISD7p")

# df of celltypes of interest (coi) in the regions of interest (roi)
df <- spatial.seurat@meta.data %>% filter(
   DetailedRegionAnnoShort %in% roi 
  ) %>% select(
    DetailedRegionAnnoShort, coi
  )

# table of mean coi per roi
mean_table <- df %>% 
  group_by(DetailedRegionAnnoShort) %>% 
  summarize(across(starts_with("RCTD_"), \(x) mean(x, na.rm = TRUE)), .groups = "drop")

# Stacked bar plot
{
  long_data <- mean_table %>% 
    pivot_longer(
      cols = starts_with("RCTD_"), 
      names_to = "CellType", 
      values_to = "MeanValue"
    ) %>% mutate(CellType = stringr::str_remove(CellType, "RCTD_"))
  
  # Create the plot
  stacked_bar_plot <- ggplot(long_data, aes(x = DetailedRegionAnnoShort, y = MeanValue, fill = CellType)) +
    geom_bar(stat = "identity", alpha = 0.9) + # Stacked bar plot
    scale_fill_manual(values = col.list$zeng.cols) +
    labs(
      x = NULL,
      y = "Average Spot Composition",
      fill = "Cell Type"
    ) + theme_mk + remove_grid

  # Display the plot
  print(stacked_bar_plot)
}
```

```{r RCTD Violin plot}
spatial.seurat %>% enh_vlnplot(
  feature = c("RCTD_Neurons", "RCTD_Astrocytes", "RCTD_Microglia", "RCTD_OLs", "RCTD_OPCs", "RCTD_PeripheralMyeloidCells"), 
  grouping = "DetailedRegionAnnoShort", 
  colors = col.list$cols_mono_short, 
  compare_means = F, idents = roi
  ) & NoLegend()
```

Correlation of cell types with genes in a time-dependent manner
```{r correlation cell types and genes}
gene_cell_correlations <-
  mapply(\(x) {
    # cell type df
    celltype_df <-
      spatial.seurat@meta.data %>%
      filter(BrainAreas %in% "Lesion" & Condition %in% x) %>%
      select(starts_with("RCTD_")) %>%
      rownames_to_column("Barcode")
    
    # expression matrix
    exp_mat <-
      spatial.seurat %>%
      GetAssayData(assay = "Spatial", layer = "data") %>%
      {
        .[rowSums(.) > (0.01 * ncol(.)),] # remove genes with less 1 % positivity
      } %>% 
      t %>%
      as.data.frame() %>%
      rownames_to_column("Barcode") %>%
      filter(Barcode %in% (celltype_df %>% pull(Barcode))) %>%
      column_to_rownames("Barcode")
    
    # correlation
    cor_matrix <-
      cor(exp_mat,
          celltype_df %>% select(-c(Barcode)),
          method = "pearson") %>%
      as.data.frame() %>% 
      rownames_to_column("GeneSymbols")
    
    return(cor_matrix)
  },
  c("1DPI", "3DPI", "7DPI"), 
  SIMPLIFY = FALSE)


result.list[["gene_cell_correlations"]] <- gene_cell_correlations
write.xlsx(result.list[["gene_cell_correlations"]], file = file.path(ws, "Correlations_gexp_CellProportion_in_lesions.xlsx"))


gene_cell_correlations[["1DPI"]] %>% 
  rownames_to_column("Gene") %>% 
  filter(Gene %in% "Ccl3")


temp <- gene_cell_correlations[["1DPI"]] %>% 
  select(RCTD_Microglia) %>% 
  filter(!is.na(RCTD_Microglia)) %>% 
  rownames_to_column("Gene") %>% 
  arrange(desc(RCTD_Microglia)) %>% 
  mutate(Percentile = row_number() / n() * 100)

# top positive and negative correlations
cc_labels <- temp %>% filter(Gene %in% c("Ccl3", "Ccl4", "Ccl12"))
top_pos <- temp %>% top_n(20, RCTD_Microglia)
top_neg <- temp %>% top_n(-5, RCTD_Microglia)
# top_genes <- bind_rows(top_pos, top_neg)

# Create the rank plot
temp_plot <- ggplot(temp, aes(x = Percentile, y = RCTD_Microglia)) +
  geom_point() +
  geom_label_repel(data = cc_labels, aes(label = Gene), 
                   color = "black", 
                   fill = col.list$zeng.cols["Microglia"], 
                   box.padding = 0.3, 
                   point.padding = 0.2, 
                   segment.color = 'grey50',
                   max.overlaps = Inf,
                   min.segment.length = 0,
                   force = 3, 
                   direction = "both", # Move labels horizontally
                   nudge_x = 25 , # Ensure labels start at least at x = 10
                   segment.size = 0.2) +
  # geom_label_repel(data = top_neg, aes(label = Gene), 
  #                  color = "black", 
  #                  fill = col.list$zeng.cols["Microglia"], 
  #                  box.padding = 0.3, 
  #                  point.padding = 0.2, 
  #                  segment.color = 'grey50',
  #                  max.overlaps = Inf,
  #                  min.segment.length = 0,
  #                  force = 1, 
  #                  direction = "both", # Move labels horizontally
  #                  nudge_x = -25, # Ensure labels end at most at x = 90
  #                  segment.size = 0.2) +
  labs(title = "Rank Plot of Correlation with RCTD_Microglia",
       x = "Gene Percentile",
       y = expression(rho[p])) +
  theme_mk + 
  remove_grid
print(temp_plot)
```

How are the cell populations related to enriched biological processes?
```{r load GO results}
result.list[["Enrichments"]] <- readRDS(file = file.path("ws", "Metascape_GO_parent_terms.Rds"))
```

Compute Module scores for the GO summary parent terms
```{r GO summaries as module scores}
# get the gene lists of GO summary parent terms
result.list[["GOsummaries"]] <- list()
for (i in result.list[["Enrichments"]][["Process_Description"]]) {
  # select the gene symbols of every process
  temp <-  result.list[["Enrichments"]] %>%
    filter(Process_Description %in% i) %>%
    pull(Symbols) %>%
    strsplit(",")
  
  # append the results
  result.list[["GOsummaries"]] %<>% append(temp)
}
names(result.list[["GOsummaries"]]) <-
  result.list[["Enrichments"]] %>% pull(Process_Description)

# turn the GO summaries into module scores.
spatial.seurat %<>% AddModuleScore(features = result.list[["GOsummaries"]],
                                   name = names(result.list[["GOsummaries"]]),
                                   assay = "SCT")

# remove environmental variables
rm(i, temp)
```

Correlate the GO Module scores with RCTD
```{r cor rctd with go parents}
rctd_df <- spatial.seurat@meta.data %>% 
  filter(BrainAreas %in% "Lesion") %>% 
  select(starts_with("RCTD_"))

go_df <- spatial.seurat@meta.data %>% 
  filter(BrainAreas %in% "Lesion") %>% 
  select(contains("DPI_"))

all(rownames(rctd_df) == rownames(go_df))

calculate_correlations <- \(df1, df2, ...) {
  
  cor_mat <- matrix(nrow = ncol(df2), ncol = ncol(df1))
  colnames(cor_mat) <- colnames(df1)
  rownames(cor_mat) <- colnames(df2)
  
  for(col1 in colnames(df1)) {
    for(col2 in colnames(df2)){
      cor_mat[col2, col1] <- cor(df1[[col1]], df2[[col2]], ...)
    }
  }
  
  # adjust for multiple hypothesis testing
  cor_mat %<>% p.adjust(method = "fdr")
  
  return(cor_mat)
}

result.list[["Cor_rctd_vs_go_parents"]] <- calculate_correlations(rctd_df, go_df, method = "spearman")

# convert the matrix to long format
cor_df <- result.list[["Cor_rctd_vs_go_parents"]] %>% as.table %>% as.data.frame()
colnames(cor_df) <- c("go_parent", "celltype", "correlation")

# plot the heatmap
cor_df %>% ggplot(aes(x = celltype, y = go_parent, fill = correlation)) + 
  geom_tile(color = "white") + 
  scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
                       midpoint = 0, limit = c(-1, 1), 
                       space = "Lab", name = "Correlation") + 
  theme_mk + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
```

Top correlated processes for each celltype
```{r top cor processes with celltypes}
temp <- cor_df %>% 
  group_by(celltype) %>% 
  arrange(desc(correlation)) %>% 
  slice_head(n = 5) %>% 
  ungroup

cor_value <- spatial.seurat@meta.data %>% 
  filter(BrainAreas %in% "Lesion") %>% 
  select(RCTD_OPCs, `7DPI_Gliogenesis53`) %>% 
  summarise(cor = cor(RCTD_OPCs, `7DPI_Gliogenesis53`)) %>%
  pull(cor)


spatial.seurat@meta.data %>%
  filter(BrainAreas %in% "Lesion") %>% 
  select(c(
    "BrainAreas", 
    "Condition", 
    "RCTD_OLs", 
    "RCTD_Astrocytes", 
    "RCTD_OPCs", 
    "RCTD_Microglia", 
    "7DPI_Gliogenesis53")) %>% 
  ggplot(aes(x = .data[["RCTD_OPCs"]], y = .data[["7DPI_Gliogenesis53"]], color = .data[["BrainAreas"]])) +
  geom_point() + 
  scale_color_manual(values = col.list$BrainAreas) +
  theme_mk + 
  remove_grid


spatial.seurat %>% FeatureScatter(
  feature1 = "RCTD_OLs", 
  feature2 = '7DPI_Gliogenesis53', 
  group.by = "BrainAreas",
  shuffle = T, 
  pt.size = 1.5, 
  cols = col.list$BrainAreas, 
  smooth = F, 
  plot.cor = T
)



```