---
title: "2V_04_Fig04_Fig05_MCAO_single_nucleus"
author: "Daniel Zucha"
date: "2023-06-07"
output: html_document
---

A markdown for the analysis of the processed single-nucleus and single-cell datasets. All .Rds can be downloaded from the Mendeley data repository.
  
  
=== Essential chunks. Always load. === 

```{r libraries}
library(Seurat)
library(SeuratWrappers)
library(dplyr)
library(tibble)
library(magrittr)
library(dittoSeq)
library(purrr)

library(gridExtra)
library(patchwork)
library(plotly)
library(pheatmap)
library(ggplot2)
library(RColorBrewer)

library(openxlsx)

options(Seurat.object.assay.version = "v3")

## sourcing helper functions
source("code/supporting_functions_MCAO.R")
```

```{r working space and lists}
if(!dir.exists("results/PNAS/Fig04")){dir.create("results/PNAS/Fig04")
  ws <- "results/PNAS/Fig04"}else{ws <- "results/PNAS/Fig04"}

# lists
seurat.plots <- list()
seurat.results <- list()
```

load datasets
```{r load seurats}
# spatial data
spatial.seurat <- readRDS(file.path("data", "seurat_spatial_1DP_06.Rds"))
sections <- c("Ctrl", "1DPI", "3DPI", "7DPI")

# single-nucleus data
sn.seurat <- readRDS(file = "data/Seurat_sn_MCAO_Zucha2023_QCFiltered.Rds")
```

== Main Figure ==

## Astroependymal populations

Loading the astro dataset
```{r astro seurat data}
# seurat
astro.seurat <- readRDS("data/seurat_ASTRO_1DP_11.rds")
```

UMAPs
```{r UMAP}
seurat.plots[["UMAP_ASTRO"]] <- astro.seurat %>%
  UMAPPlot(
    group.by = "cell_type_2",
    pt.size = 0.25,
    label = F,
    label.box = T, 
    repel = T,
    cols = col.list[["astro_cols"]]
  ) +
  NoLegend() +
  theme_mk +
  remove_grid +
  NoAxes() +
  theme(panel.border = element_blank(), 
        legend.position = "none") + 
  labs(title = NULL)

print(seurat.plots[["UMAP_ASTRO"]])

# save 
  ggsave(filename = file.path(ws, "/UMAP_sn_Astroepe.tiff"), device = 'tiff', dpi = 300, units = 'cm', width = 10, height = 10)
```

Population proportions clusters - condition
```{r Proportions}
## proportions
library(dittoSeq)
astro.seurat$Condition %>% table
astro.seurat$cell_type_2 %>% table

astro.seurat %>% dittoBarPlot(var = "Condition", group.by = "cell_type_2", scale = "count")
astro.seurat %>% dittoBarPlot(var = "Condition", group.by = "cell_type_2", scale = "percent")

## data frames
astro.seurat$Condition %<>% factor(levels = c("sham", "D1", "D3", "D7"))

calculate_proportions <- \(seurat, group.by, celltype_annotation) {
  # initialize the proportion matrix
  ids <- seurat@meta.data[[group.by]] %>% levels
  categories <- seurat@meta.data[[celltype_annotation]] %>% levels
  counts <- matrix(nrow = length(ids), ncol = length(categories))
  rownames(counts) <- ids
  colnames(counts) <- categories
  
  # count the occurences
  for (i in seq_along(ids)) {
    for (j in seq_along(categories)) {
      count <- seurat@meta.data %>%
        filter(.data[[group.by]] == ids[i], .data[[celltype_annotation]] == categories[j]) %>%
        nrow()
      
      counts[i, j] <- count
    }
  }
  
  
  # get the percent count proportions
  require(tidyr)
  counts <- counts/rowSums(counts)
  counts_mm_1 <- counts %>% 
    as.data.frame %>% 
    rownames_to_column(group.by) %>% 
    tidyr::pivot_longer(
      cols = -all_of(group.by),
      names_to = celltype_annotation, 
      values_to = "value")
    
  return(counts_mm_1)
}

seurat.results[["sn_astro_proportions"]] <- 
  calculate_proportions(
    seurat = astro.seurat, 
    group.by = "Condition", 
    celltype_annotation = "cell_type_2")

ggplot(data = seurat.results[["sn_astro_proportions"]], aes(x = cell_type_2, y = value, fill = Condition)) + ## percent count proportions
  geom_col(position = "dodge") +
  theme_classic() +
  theme(
    axis.text.x = element_text(
      angle = 105,
      hjust = 1,
      vjust = 0,
      size = 35,
      color = "black"
    ),
    axis.text.y = element_text(
      angle = 90,
      hjust = 0.5,
      vjust = 0.5,
      size = 18,
      color = "black"
    ),
    axis.title = element_text(family = "OpenSans", size = 25),
    plot.background = element_blank(),
    panel.background = element_blank()
  ) +
  ylab('Proportion of nuclei') +
  xlab("") + ggtitle("") + NoLegend() + ylim(0, (max(seurat.results[["sn_astro_proportions"]]$value) + 0.1)) +
  scale_fill_manual(values = col.list$Condition)


p %>% ggsave(filename = paste0(ws, "/BarPlot_PercentProportions_sn_astro_bySample.tiff"), device = 'tiff', units = "cm", dpi = 300, width = 10, height = 8, bg = 'transparent')
```

Markers and dotplot
```{r Markers}
all.markers <- FindAllMarkers(astro.seurat, logfc.threshold = 0.58, assay = 'SCT', verbose = T, only.pos = T) %>% 
  arrange(desc(avg_log2FC)) %>% 
  split(f = .[["cluster"]]); names(all.markers) <- paste0("SN_Ast_", names(all.markers))
write.xlsx(all.markers, file = file.path(ws, "Markers_sn_AstroEpe.xlsx"), overwrite = T)

astro.seurat %>% FeaturePlot(features = c("Dcx", "Foxj1", "Tmem72"), label = T)

## dotplot
astro.seurat$cell_type_2 %>% levels

features_test <- c('Dcx', 'Nrxn3', "Adarb2",   # Neuroblasts
                 'Foxj1', 'Adamts20', 'Tmem212', # Ependymal
                 'Tmem72', 'Htr2c', "Otx2os1",  # Choroid Plexus cells
                 'Grm3', 'Rgs7', 'Htra1',  # TE Astro
                 'Slc1a3', "Aldh1l1",  # astro general
                 "Gria1", "Fry", "Slc6a11", "Mgat4c",  # DE Astro
                 'Ogt', 'Fam107a', 'Cox8a', 'Aldoc', "Shank1", "Nrgn", # Activated Astro
                 "Ctsd", 'Gfap', "Clu", 'Apoe', "Ptn", "Gpm6b", "Cd9", "Vim", "Thbs4"    # Reactive astro
                 )

p <-
  DotPlot(
    astro.seurat,
    assay = "SCT",
    features = features_test,
    group.by = "cell_type_2",
    cols = c("#FFF7FB", "#014636")
  ) +
  labs(x = NULL, y = NULL) +
  theme_mk +
  remove_grid + 
  theme(axis.text.x = element_text(angle = 55, vjust = 1, hjust = 1), 
        axis.ticks = element_blank())

png_save_show(
  plot = p, 
  file = file.path(ws, "DotPlot_sn_Astroepe.png"), 
  show_plot = FALSE, 
  dpi = 1000,
  height = 80, 
  width = 200)
```

Mitochondrial and ribosomal content
```{r mito and ribo}
astro.seurat[["percent_mt"]] <- PercentageFeatureSet(astro.seurat, pattern = "mt-")
astro.seurat %>% FeaturePlot(features = "percent_mt", pt.size = 0.25, max.cutoff = 'q99', cols = brewer.pal(n = 9, name = "Blues")[3:9]) %>% 
  ggsave(filename = file.path(ws, "FeaturePlot_sn_astro_percentMT.tiff"), device = 'tiff', dpi = 300, width = 10, height = 10, units = 'cm')

astro.seurat[["percent_rib"]] <- PercentageFeatureSet(astro.seurat, pattern = "Rpl|Rps")
astro.seurat %>% FeaturePlot(features = "percent_rib", pt.size = 0.25, max.cutoff = 'q99', cols = brewer.pal(n = 9, name = "Blues")[3:9]) %>% 
  ggsave(filename = file.path(ws, "FeaturePlot_sn_astro_percentRIB.tiff"), device = 'tiff', dpi = 300, width = 10, height = 10, units = 'cm')
```

Gene ontology analysis
```{r GO analysis}
astro.seurat$cell_type_2 %>% levels

## GO analysis - Reac Astro vs DE Astro ####
diff.reac.markers <- astro.seurat %>% FindMarkers(ident.1 = "Reactive Astrocytes", ident.2 = c("DE Astrocytes"), logfc.threshold = 0.58, only.pos = T, assay = 'SCT', verbose = T) %>%
  arrange(desc(avg_log2FC)) %>% 
  filter(p_val_adj < 0.01)

reac.markers <- list(
  "Reactive Astrocytes" = diff.reac.markers %>% filter(avg_log2FC > 0.58) %>% rownames() %>% .[grep("^Rp|^mt", x = ., ignore.case = F, invert = T)]
)

library(clusterProfiler); library("org.Mm.eg.db", character.only = TRUE)
organism <- org.Mm.eg.db; background <- rownames(astro.seurat[["SCT"]])
ORA.list <- lapply(reac.markers, function(x) enrichGO(gene          = x, # list of genes 
                                                      universe      = background, # 'potential DEGs'
                                                      keyType       = "SYMBOL", # or ENSEMBL
                                                      OrgDb         = organism,
                                                      ont           = "ALL", # BP = biological process, MF = molecular function, CC = cellular component
                                                      pAdjustMethod = "fdr",
                                                      pvalueCutoff  = 0.01, #
                                                      qvalueCutoff  = 0.2,
                                                      readable      = F,
                                                      minGSSize     = 5) %>% clusterProfiler::simplify()
                   )
ORA.list %>% write.xlsx(file = file.path(ws, "ORAlist_sn_ReacAstro_vs_DEastro.xlsx"), overwrite = T)

## selecting terms for the barplot
ORA.list <- read_all_sheets(file = file.path(ws, "ORAlist_sn_ReacAstro_vs_DEastro.xlsx"))
terms <- c(
           ## BP
           "GO:0007263", # NO mediated signal transduction
           "GO:0014009", # glial cell proliferation
           "GO:0050767", # regulation of neurogenesis
           "GO:0040013", # negative regulation of locomotion
           "GO:0061687", # detoxification of inorganic compound
           
           ## CC
           "GO:0005764", # lysosome
           "GO:0045121", # membrane raft
           "GO:1990777", # lipoprotein particle
           "GO:0030312", # external encapsulating structure
           
           ## MF
           "GO:0050839" # cell adhesion molecule binding
           )

{
  population <- "Reactive Astrocytes"
  subset.ORA <- ORA.list[[population]] %>% filter(ID %in% terms)
  ORA.data <- data.frame(
    'Description' = subset.ORA[["Description"]] %>% stringr::str_to_title(),
    'Pval' = subset.ORA[["p.adjust"]] %>% log10 %>% '*'(-1))
  ORA.data$Description %<>% factor(levels = rev(unique(.)))
  rm(subset.ORA)
  
  (ggplot(data = ORA.data, aes(x = Description, y = Pval)) + 
    geom_bar(stat = "identity", width = 0.6, size = 0.2, colour = "black", alpha = 0.75, fill = col.list[["astro_cols"]][[population]]) +
    ylim(0, max(ORA.data$Pval) + 1) + 
    geom_hline(yintercept = 1.3, size = 0.2) +
    coord_flip() +
    theme(text = element_text(family = "OpenSans", size = 40), 
          axis.text.x = element_text(vjust = 1, size = 40),
          axis.text.y = element_text(vjust = 0.5, size = 40, color = 'black'),
          axis.ticks = element_blank(),
          panel.background = element_blank(), #transparent panel bg
          plot.background = element_blank(), #transparent plot bg 
          panel.grid.major.x = element_blank(),
          panel.grid.major.y = element_line(colour = "#A1A19C", size = 0.2)
          ) +  
          xlab("") + ylab(expression(paste("-log(Padj)")))) %>% ggsave(filename = paste0(ws, "/Barplot_sn_ReacAstro_GO_", population, ".tiff"), device = "tiff", dpi = 300, width = 7, height = 5)
}

## GO analysis - Activated vs DE Astro ####
diff.reac.markers <- astro.seurat %>% FindMarkers(ident.1 = "Activated Astrocytes", ident.2 = c("DE Astrocytes"), logfc.threshold = 0.58, only.pos = T, assay = 'SCT', verbose = T) %>%
  arrange(desc(avg_log2FC)) %>% 
  filter(p_val_adj < 0.01)

reac.markers <- list(
  "Activated Astrocytes" = diff.reac.markers %>% filter(avg_log2FC > 0.58) %>% rownames() %>% .[grep("^Rp|^mt", x = ., ignore.case = F, invert = T)]
)

library(clusterProfiler); library("org.Mm.eg.db", character.only = TRUE)
organism <- org.Mm.eg.db; background <- rownames(astro.seurat[["SCT"]])
ORA.list <- lapply(reac.markers, function(x) enrichGO(gene          = x, # list of genes 
                                                      universe      = background, # 'potential DEGs'
                                                      keyType       = "SYMBOL", # or ENSEMBL
                                                      OrgDb         = organism,
                                                      ont           = "ALL", # BP = biological process, MF = molecular function, CC = cellular component
                                                      pAdjustMethod = "fdr",
                                                      pvalueCutoff  = 0.01, #
                                                      qvalueCutoff  = 0.2,
                                                      readable      = F,
                                                      minGSSize     = 5) %>% clusterProfiler::simplify()
                   )
ORA.list %>% write.xlsx(file = file.path(ws, "ORAlist_sn_ActivatedAstro_vs_DEastro.xlsx"), overwrite = T)
```

Projections on how the origin of the astro populations, based on the Zeisel et al annotation. Calculation of the projections is in a separate markdown.
```{r Brain region projections}
## plot
astro.seurat@assays %>% names
DefaultAssay(astro.seurat) <- "prediction_ASTRO_region" ## azimuth projection based on Zeisel et al
astro.seurat@assays$prediction_ASTRO_region %>% dimnames %>% .[1] ## brain region names


## where are the reactive astro populations from?
astro.seurat %>% VlnPlot(features = "Telencephalon", group.by = "cell_type_2") ## telencephalon
astro.seurat %>% VlnPlot(features = "Hypothalamus,Thalamus,Midbrain dorsal,Midbrain ventral,Pons,Medulla,Spinal cord", group.by = "cell_type_2") ##diencephalon

(FeaturePlot(astro.seurat, features = c("Telencephalon"), pt.size = 0.25, cols = brewer.pal(n = 9, name = "BuGn")[3:9]) + ## telencephalon projection
  theme(plot.background = element_blank(), panel.background = element_blank())) %>% 
  ggsave(filename = file.path(ws, "FeaturePlot_sn_Astroepe_Telencephalon_projection.tiff"), device = 'tiff', units = "cm", dpi = 300, width = 10, height = 10)

(FeaturePlot(astro.seurat, features = c("Hypothalamus,Thalamus,Midbrain dorsal,Midbrain ventral,Pons,Medulla,Spinal cord"), pt.size = 0.25, cols = brewer.pal(n = 9, name = "BuGn")[3:9]) + ## diencephalon projection
  theme(plot.background = element_blank(), panel.background = element_blank())) %>% 
  ggsave(filename = file.path(ws, "FeaturePlot_sn_Astroepe_Diencephalon_projection.tiff"), device = 'tiff', units = "cm", dpi = 300, width = 10, height = 10)

```

Visualising the astroependymal populations in the spatial data.
```{r Astro in Spatial}
astro.markers <- read_all_sheets(file.path(ws, "Markers_sn_AstroEpe.xlsx"))

## add to spatial data
top.astro.markers <- lapply(names(astro.markers), function(astro){glist <- astro.markers[[astro]][["gene"]] %>% head(15); return(glist)}); names(top.astro.markers) <- names(astro.markers)
spatial.seurat %<>% AddModuleScore(features = top.astro.markers, name = names(top.astro.markers))
spatial.seurat[[]] %>% names %>% tail(10) ## module scores were added

top.astro.markers %>% names %>% length() ## 7 astroepe populations 
features <- spatial.seurat[[]] %>% names %>% tail(7)

## plot the spatial plot
lapply(features, function(feature){
  feature.name <- feature %>% stringr::str_sub(end = -2)
  print(feature.name)
  minmax <- c(min(spatial.seurat[[feature]]), max(spatial.seurat[[feature]]))
  (spatial.seurat %>% 
      SpatialPlot(features = feature, crop = F, pt.size = 1.35, alpha = c(0.8, 1), images = c('sham', 'D1', 'D3', 'D7'), image.alpha = 0, stroke = 0.01) & 
      theme(text = element_text(family = plot.font, size = 9), legend.key.size = unit(0.2, 'cm')) #& NoLegend() 
    & ggplot2::scale_fill_gradientn(colours = c("#4D4D4D", "#FFFFFF", "#016C59"), limits = minmax)
    ) %>% 
  ggsave(filename = paste0(ws, "/SpatialPlot_snAstro_", feature.name, ".tiff"), device = 'tiff', dpi = 300, width = 18, height = 6, units = 'cm')
})
```

Looking at the published datasets and overlaps with our reactive population.
```{r metanalysis}
DefaultAssay(astro.seurat) <- "SCT"
astro.markers <- read_all_sheets(file.path(ws, "Markers_sn_AstroEpe.xlsx"))
metadata.list <- read_all_sheets(file = "data/SupplementaryTable4_Metanalysis_GlialReactivityMarkers.xlsx") %>% 
  .[names(.) %>% grep(pattern = "astro|DAA", ignore.case = T)] ## keep the astro-related populations

## combined UMAP and VlnPlot
lapply(features, function(features){
    minmax <-  c(astro.seurat[[features]] %>% min, astro.seurat[[features]] %>% max)
    p1 <- astro.seurat %>% FeaturePlot(features = features, max.cutoff = 'q99', pt.size = 1, cols = brewer.pal(n = 9, name = "BuGn")[3:9]) + NoLegend() +
      theme(text = element_blank(), 
            plot.margin = unit(c(0,0,0,0), "cm"), 
            plot.background = element_blank(), 
            panel.background = element_blank())
    
    p2 <- VlnPlot(astro.seurat, features = features, pt.size = 0, group.by = "cell_type_2", cols = col.list[["astro_cols"]]) + coord_flip() & scale_y_continuous(breaks = scales::breaks_pretty(n = 3)) & xlab("") & labs(title = "") & ylim(minmax) &
    theme(axis.text.x = element_text(family = 'OpenSans', angle = 30, size = 10, hjust = 0.7, vjust = 0.8), 
          axis.text.y = element_text(family = 'OpenSans', size = 10),
          text = element_text(family = "OpenSans", size = 25), 
          legend.position = "none", 
          plot.margin = unit(c(0, 0.3, 0.3, 0), "cm"),
          plot.background = element_blank(), 
          panel.background = element_blank())
    feature.name <- features %>% stringr::str_sub(end = -2)
    p <- gridExtra::grid.arrange(p1, p2, ncol = 2, widths = c(1,0.5)) %>% ggsave(filename = paste0(ws, "/FeaturePlot_VlnPlot_Metanalysis_sn_Astroepe_", feature.name, ".tiff"), device = 'tiff', dpi = 300, bg = NULL, width = 9, height = 6)
})

## overlap testing
astro.markers %>% names

## only positive markers loaded?
(astro.markers[["SN_Ast_Reactive Astrocytes"]][["avg_log2FC"]] %>% min) > 0.58
rastro.markers <- astro.markers[["SN_Ast_Reactive Astrocytes"]][["gene"]]

meta.intersects <- lapply(names(metadata.list), function(meta){
  paste0("Length of metadataset ", meta, " is ",  metadata.list[[meta]][[1]] %>% length, " genes.") %>% print
  int <- intersect(rastro.markers, metadata.list[[meta]][[1]]) %>% length
  paste0("Intersection is ", int) %>% print
  return(int)
  }); names(meta.intersects) <- names(metadata.list)

## signif testing gene overlaps
### DAA and our reac ####
  
meta.signif <- lapply(names(meta.intersects), function(meta){
  subset.reac.astro <- astro.seurat %>% subset(subset = cell_type_2 == "Reactive Astrocytes")
  background.size <- ((subset.reac.astro@assays$RNA %>% rowSums()) > 10) %>% sum ##background gene set size with UMI counts > 10
  intersection <- meta.intersects[[meta]]
  meta.set.size <- metadata.list[[meta]] %>% length
  phyper(
  q = intersection - 1,  ## number of red marbles in the draw -1 (see below)
  m = meta.set.size, ## number of red marbles in urn
  n = background.size - meta.set.size, ## number of green marbles in urn
  k = rastro.markers %>% length, ## number of drawn marbles
  lower.tail = FALSE ## compute P(X > overlap), hence the '-1' in the q argument
) -> test
  return(test)
}); names(meta.signif) <- names(meta.intersects); meta.signif %<>% p.adjust(method = "fdr")
meta.signif; meta.signif < 0.01  ## see results; all below

## Intersecting genes
int.genes <- intersect(rastro.markers, metadata.list[["Habib2020_DAA_5XFAD"]][[1]])

## Non-intersecting genes
subset.reac.astro <- astro.seurat %>% subset(subset = cell_type_2 == "Reactive Astrocytes")
background <- rownames(subset.reac.astro@assays$RNA)[((subset.reac.astro@assays$RNA %>% rowSums()) > 10)] ## background gene set of our reactive astrocytes.

## Our reactive markers not found in DAA
rastro.markers.unique <- rastro.markers[!(rastro.markers %in% int.genes)] 

## DAA markers not found in our reactive astros
DAA.markers.unique <- metadata.list[["Habib2020_DAA_5XFAD"]][[1]][!(metadata.list[["Habib2020_DAA_5XFAD"]][[1]] %in% int.genes)]

## Gene Ontology of the non-overlapping DAA markers. = What processes are missing in our reactive astrocytes?
library(clusterProfiler); library("org.Mm.eg.db", character.only = TRUE); organism <- org.Mm.eg.db
GO.DAA.markers.unique <- enrichGO(gene          = DAA.markers.unique, # list of genes 
                                  universe      = background, # 'potential DEGs'
                                  keyType       = "SYMBOL", # or ENSEMBL
                                  OrgDb         = organism,
                                  ont           = "ALL", # BP = biological process, MF = molecular function, CC = cellular component
                                  pAdjustMethod = "fdr",
                                  pvalueCutoff  = 0.01, 
                                  qvalueCutoff  = 0.2,
                                  readable      = F,
                                  minGSSize     = 5) #%>% clusterProfiler::simplify()




## Published datasets in spatial
## add metadata
spatial.seurat %<>% AddModuleScore(features = metadata.list, name = metadata.list %>% names)
features <- names(metadata.list) %>% paste0(., seq(1:length(.))) ## feature names as they are added
identical(features, spatial.seurat[[]] %>% names %>% tail(length(features))) ## TRUE = vector names are identical

## correlation our Reactive and DAA Habib
spatial.seurat %<>% AddModuleScore(features = list(rastro.markers %>% head(15)), name = "snAst_Reac")

spatial.seurat[[]] %>% names
df <- data.frame(
  "snAst_Reac" = spatial.seurat$snAst_Reac1,
  "Habib_DAA" = spatial.seurat$Habib2020_DAA_vs_homeo2
)

cor.test(x = df$snAst_Reac, y = df$Habib_DAA, method = "spearman")
plot(x = df$snAst_Reac, y = df$Habib_DAA)

(ggplot(data = df, aes(x = snAst_Reac, y = Habib_DAA)) + 
  geom_point(aes(color = snAst_Reac), size = 0.5) + 
    scale_color_gradientn(colors = c("#3690C0", "#014636")) +
  geom_smooth(method = 'lm', se = F, color = "black") + xlab("Reactive Astrocyte signature") + ylab("DAA signature") + theme_light() +
    theme(axis.title = element_text(size = 20, family = "OpenSans"))) %>% ggsave(filename = file.path(ws, "ScatterPlot_snAstroReac_HabibDAA.tiff"), device = 'tiff', dpi = 300, height = 5, width = 7, units = "cm")

## signif testing on module scores
features
Astrosplitlist <- astro.seurat@meta.data %>% group_split(cell_type_2); names(Astrosplitlist) <- astro.seurat$cell_type_2 %>% levels
wilcox.meta <- lapply(features, function(feature){
  temp <- lapply(c('Reactive_Astro1', 'Reactive_Astro2'), function(AS){
    test.res <- wilcox.test(x = Astrosplitlist[[AS]][[feature]],
                            y = c(Astrosplitlist[['ASTRO_DE']][[feature]], Astrosplitlist[['ASTRO_TE']][[feature]]), conf.int = T)
    test.pval <- test.res[["p.value"]]
  }); names(temp) <- c('Reactive_Astro1', 'Reactive_Astro2')
  return(temp)
}); names(wilcox.meta) <- features; wilcox.meta %<>% unlist %>% p.adjust(method = 'fdr'); wilcox.meta %>% .[. < 0.01] ## adjusting p values for multiple hypothesis testing and selecting only those with pval< 0.01

wilcox.meta[wilcox.meta < 0.0001] %>% names

```

Immunovalidation Aldh1l1 + Gfap
```{r immuno astro}
immuno <- 
  read_all_sheets(
    file = file.path("results", "PNAS", "Tables", "MCAO_IHC_Validations.xlsx")
    ) %>% 
  .[["Aldh1l1_Gfap"]] %>% 
  mutate(Proportion_Colocalized = Proportion_Colocalized * 100) %>% 
  dplyr::mutate(Condition = factor(Condition, levels = sections)) %>% 
  group_by(SampleName, Condition) %>% 
  dplyr::summarise(
    mean_colocalization = mean(Proportion_Colocalized), 
    .groups = "drop"
  )

immuno_summary <- 
  immuno %>% 
  group_by(Condition) %>% 
  dplyr::summarise(
    sd_colocalization = stats::sd(mean_colocalization),
    mean_colocalization = mean(mean_colocalization)
  )

# stat test control vs injured
stat_test <- 
  immuno %>% 
  mutate(Injured =  if_else(Condition == "Ctrl", "NoMCAO", "MCAO")) %>% 
  wilcox.test(mean_colocalization ~ Injured, data = .)
print(stat_test)




# plot
seurat.plots[["Barplot_Immuno_Astro"]] <- 
  ggplot(
    data = immuno, 
    aes(y = mean_colocalization, x = Condition)
    ) + 
  geom_bar(data = immuno_summary, stat = "identity", width = 0.6, colour = "black", alpha = 0.75, fill = "#014636", size = 0.5) +
  geom_errorbar(
    data =  immuno_summary, 
    aes(ymin = mean_colocalization - sd_colocalization, 
        ymax = mean_colocalization + sd_colocalization), width = 0.1, color = "black", size = 0.5) +
  geom_jitter(position = position_jitter(0.2), color = 'black', size= 1.5, shape = 1, stroke = 0.5) + 
  annotate(geom = "text", x = 1, y = 40, label = paste0("P = ", stat_test$p.value %>% round(3)), vjust = -0.5, hjust = 0, size = 2.45) +
  ylim(-1, 60) +
  xlab(NULL) + 
  ylab("ALDH1L1 + GFAP co-localization (%)") +
  theme_mk + 
  remove_grid + 
  theme(panel.grid.major.y = element_line(linewidth = 0.1, colour = "grey"))

print(seurat.plots[["Barplot_Immuno_Astro"]])

png_save_show(
  plot = seurat.plots[["Barplot_Immuno_Astro"]], 
  file = file.path(ws, "BarPlot_astrocytes_Immuno.png"), 
  show_plot = FALSE, 
  dpi = 1000, 
  height = 70, 
  width = 50
)

```


## Microglial populations

Loading the seurat and making a separate folder for microglia results.
```{r seurat load}
mg.seurat <- readRDS("data/seurat_MG_names_soupX.rds"); DefaultAssay(mg.seurat) <- "SCT"

mg.seurat$cell_type_2 %>% levels
mg.seurat %<>% subset(idents = "debris", invert = T)  ## remove dead cells
mg.seurat$cell_type_2 <- case_when(
  mg.seurat$cell_type_2 == "homeostatic" ~ "Homeostatic",
  mg.seurat$cell_type_2 == "TAM" ~ "Reactive"
); mg.seurat$cell_type_2 %<>% factor(levels = c("Homeostatic", "Reactive")); mg.seurat %<>% SetIdent(value = "cell_type_2")
mg.seurat[[]] %>% names() ## metadata overview

col.list[["mg_cols"]] <- c(
  "Homeostatic" = "#FC9272",
  "Reactive" = "#A50F15"
)
```

UMAP
```{r Figure sn UMAP}
(mg.seurat %>% 
  UMAPPlot(group.by = "cell_type_2", pt.size = 1, label = F, label.box = T, cols = col.list[["mg_cols"]]) + 
  NoLegend() + 
  theme(plot.background = element_blank(), panel.background = element_blank())) %>% ggsave(filename = file.path(ws, "UMAPPlot_sn_mg.tiff"), device = 'tiff', dpi = 300, units = 'cm', width = 10, height = 10)
```

Proportions per conditions.
```{r Figure sn Proportions}
## proportions
library(dittoSeq)
mg.seurat$Condition %>% table
mg.seurat$cell_type_2 %>% table

mg.seurat %>% dittoBarPlot(var = "Condition", group.by = "cell_type_2", scale = "count")
mg.seurat %>% dittoBarPlot(var = "Condition", group.by = "cell_type_2", scale = "percent")

## data frames
ids <- mg.seurat@meta.data$Condition %>% levels; categories <- unique(mg.seurat@meta.data$cell_type_2); counts <- matrix(nrow=length(ids), ncol=length(categories)); rownames(counts) <- ids; colnames(counts) <- categories
for (i in seq_along(ids)) {
  for (j in seq_along(categories)) {
    count <- mg.seurat@meta.data %>%
      filter(Condition == ids[i], cell_type_2 == categories[j]) %>%
      nrow()

    counts[i, j] <- count
  }
}; rm(i,j)

counts_mm_0 <- reshape2::melt(counts, id = "rownames") ## total count proportions

ggplot(data = counts_mm_0, aes(x = Var2, y = value, fill = Var1)) + 
  geom_col(position = "dodge") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  ylab('Proportion of cells')
counts <- counts/rowSums(counts)
counts_mm_1 <- reshape2::melt(counts, id = "rownames") ## percent count proportions


p <-  ggplot(data = counts_mm_1, aes(x = Var2, y = value, fill = Var1)) + ## percent count proportions
    geom_col(position = "dodge") +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 105, hjust = 1, vjust = 0, family = "OpenSans", size = 35, color = "black"), 
          axis.text.y = element_text(angle = 90, hjust = 0.5, vjust = 0.5, family = "OpenSans", size = 18, color = "black"),
          axis.title = element_text(family = "OpenSans", size = 25),
          plot.background = element_blank(),
          panel.background = element_blank()) +
    ylab('Proportion of nuclei') +
    xlab("") + ggtitle("") + NoLegend() + ylim(0,1) +
    scale_fill_manual(values = c("#525475", "#F1D7AB", "#E6B188", "#D18080")); p
p %>% ggsave(filename = file.path(ws, "BarPlot_PercentProportions_sn_MG_bySample.tiff"), device = 'tiff', units = "cm", dpi = 300, width = 6, height = 8, bg = 'transparent')
```

Marker genes, dotplot, volcano plot.
```{r Markers and Volcano}
## Markers
mg.seurat.markers.TAM <- FindMarkers(mg.seurat, ident.1 = "Reactive", ident.2 = "Homeostatic", assay = "SCT", logfc.threshold = 0.58, only.pos = F, verbose = T) %>% arrange(desc(avg_log2FC)) %>% filter(p_val_adj < 0.01)

mg.seurat.markers <- FindAllMarkers(object = mg.seurat, assay = "SCT", logfc.threshold = 0.58, verbose = T, only.pos = T) %>% 
  arrange(desc(avg_log2FC)) %>% 
  filter(p_val_adj < 0.01) %>% 
  split(f = .[["cluster"]]); names(mg.seurat.markers) <- paste0("SN_MG_", names(mg.seurat.markers))

mg.seurat.markers %>% write.xlsx(file = file.path(ws, "Markers_sn_MG.xlsx"), overwrite = T)
mg.seurat.markers <- read_all_sheets(file = file.path(ws, "Markers_sn_MG.xlsx"))

## Dotplot
features_test <- c('Hexb', 'Cx3cr1', 'P2ry12', 'Tmem119',  # homeo MG
                   'Apoe', 'Ctsb', 'Abca1',  'Hif1a', 'Spp1', 'Sash1',  'H2-D1', 'Lyz2', 'Trem2', 'Igf1', "Lgals3",  'Gpnmb', 'Aif1'  ## reactive mg
                     )

p <-
  DotPlot(
    mg.seurat,
    assay = "SCT",
    features = features_test,
    group.by = "cell_type_2",
    cols = c("#FFF7FB", "#A50F15")
  ) +
  labs(x = NULL, y = NULL) +
  theme_mk +
  remove_grid + 
  theme(axis.text.x = element_text(angle = 55, vjust = 1, hjust = 1), 
        axis.ticks = element_blank())

png_save_show(
  plot = p, 
  file = file.path(ws, "DotPlot_sn_microglia_markers.png"), 
  show_plot = FALSE, 
  dpi = 1000,
  height = 60, 
  width = 150)

  
## Volcano plot ####
library(EnhancedVolcano)
marker.df <- mg.seurat.markers.TAM
keyvals.colour <- ifelse(
  marker.df$avg_log2FC < -0.58 & marker.df$p_val_adj < 0.01, '#3288BD',
  ifelse(marker.df$avg_log2FC > 0.58 & marker.df$p_val_adj < 0.01, '#D53E4F',
         '#E7E7E7')) 
names(keyvals.colour)[keyvals.colour == '#3288BD'] <- 'downregulated'; names(keyvals.colour)[keyvals.colour == '#D53E4F'] <- 'upregulated'; names(keyvals.colour)[keyvals.colour == '#E7E7E7'] <- 'insignificant'

p <- EnhancedVolcano(toptable = marker.df,
                     lab = rownames(marker.df),
                     x = 'avg_log2FC',
                     y = 'p_val_adj',
                     axisLabSize = 20,
                     titleLabSize = 14,
                     title = "",
                     pCutoff = 0.05,
                     colAlpha = 0.7,
                     colCustom = keyvals.colour,
                     subtitleLabSize = 9,
                     FCcutoff = 0.4,
                     labSize = 10,
                     pointSize = 3.0,
                     drawConnectors = T,
                     max.overlaps = 10,
                     maxoverlapsConnectors = NULL,
                     arrowheads = F,
                     boxedLabels = T,
                     selectLab = c('P2ry12', 'Cx3cr1', 'Lyz2', 'Gpnmb', 'Apoe', 'Trem2', 'Spp1', 'Igf1', 'Ctsb', 'Lpl', 'Sash1', 'H2-D1'), ## select genes to highlight
                     labFace = "bold",
                     cutoffLineType = "blank") + 
  theme_light() + theme(axis.text = element_text(size = 30), axis.title = element_text(size = 30)) +
  NoLegend(); p
p %>% ggsave(filename = file.path(ws, "VolcanoPlot_sn_MG_ReactiveVsOthers.tiff"), device = 'tiff', units = "cm", dpi = 300, width = 12, height = 12)
```

Gene ontology analysis
```{r GO analysis}
mg.seurat$cell_type_2 %>% levels
mg.seurat.markers <- read_all_sheets(file.path(ws, "Markers_sn_MG.xlsx"))
mg.seurat.markers[["Reactive"]]$avg_log2FC %>% min ## min value, need to filter?
mg.seurat.markers[["Reactive"]]$p_val_adj %>% max ## max value, need to filter?

diff.reac.list <- list(
  "Reactive" = mg.seurat.markers[["Reactive"]] %>% filter(avg_log2FC > 0.58) %>% pull(gene) %>% .[grep("^Rp|^mt", x = ., ignore.case = F, invert = T)]
)

library(clusterProfiler); library("org.Mm.eg.db", character.only = TRUE)
organism <- org.Mm.eg.db; background <- rownames(mg.seurat[["SCT"]])
ORA.list <- lapply(diff.reac.list, function(x) enrichGO(gene          = x, # list of genes 
                                                      universe      = background, # 'potential DEGs'
                                                      keyType       = "SYMBOL", # or ENSEMBL
                                                      OrgDb         = organism,
                                                      ont           = "ALL", # BP = biological process, MF = molecular function, CC = cellular component
                                                      pAdjustMethod = "fdr",
                                                      pvalueCutoff  = 0.01, #
                                                      qvalueCutoff  = 0.2,
                                                      readable      = F,
                                                      minGSSize     = 5) %>% clusterProfiler::simplify()
                   )

ORA.list %>% write.xlsx(file = file.path(ws, "ORAlist_sn_ourReactive_vs_Homeo.xlsx"), overwrite = T)

## selecting terms for the barplot
terms <- c(
           ## BP
           "GO:1905952", # regulation of lipid localization
           "GO:0030335", # positive regulation of cell migration
           "GO:0048771", # tissue remodeling
           "GO:1903530", # regulation of secretion by cell
           "GO:0050670", # regulation of lymphocyte proliferation
           
           ## CC
           "GO:0030312", # external encapsulating structure
           "GO:0030055", # cell-substrate junction
           
           
           ## MF
           "GO:0005178", # integrin binding
           "GO:0071813", # lipoprotein particle binding
           "GO:0050839" # cell adhesion molecule binding
           )

{
  population <- "Reactive"
  subset.ORA <- ORA.list[[population]]@result %>% filter(ID %in% terms)
  ORA.data <- data.frame(
    'Description' = subset.ORA[["Description"]] %>% stringr::str_to_title(),
    'Pval' = subset.ORA[["p.adjust"]] %>% log10 %>% '*'(-1))
  ORA.data$Description %<>% factor(levels = rev(unique(.)))
  rm(subset.ORA)
  
  (ggplot(data = ORA.data, aes(x = Description, y = Pval)) + 
    geom_bar(stat = "identity", width = 0.6, size = 0.2, colour = "black", alpha = 0.75, fill = col.list[["mg_cols"]][[population]]) +
    ylim(0, max(ORA.data$Pval) + 1) + 
    geom_hline(yintercept = 1.3, size = 0.2) +
    coord_flip() +
    theme(text = element_text(family = "OpenSans", size = 40), 
          axis.text.x = element_text(vjust = 1, size = 40),
          axis.text.y = element_text(vjust = 0.5, size = 40, color = 'black'),
          axis.ticks = element_blank(),
          panel.background = element_blank(), #transparent panel bg
          plot.background = element_blank(), #transparent plot bg 
          panel.grid.major.x = element_blank(),
          panel.grid.major.y = element_line(colour = "#A1A19C", size = 0.2)
          ) +  
          xlab("") + ylab(expression(paste("-log(Padj)")))) %>% ggsave(filename = paste0(ws, "/Barplot_sn_mg_GO_", population, ".tiff"), device = "tiff", dpi = 300, width = 7, height = 5)
}
```

Single-nucleus signature in the spatial data.
```{r Spatial plot of sn signatures}
mg.seurat.markers <- read_all_sheets(file.path(ws, "Markers_sn_MG.xlsx"))

## add to spatial data
mg.names <- paste0("snMG_", names(mg.seurat.markers)); top.mg.markers <- lapply(names(mg.seurat.markers), function(mg){glist <- mg.seurat.markers[[mg]][["gene"]] %>% head(15); return(glist)}); names(top.mg.markers) <- names(mg.seurat.markers)
spatial.seurat %<>% AddModuleScore(features = top.mg.markers, name = mg.names)
spatial.seurat[[]] %>% names %>% tail ## last two are our sn mg populations

## plot the spatial plot alone
lapply(c("snMG_homeostatic1", "snMG_Reactive2"), function(feature){
  feature.name <- feature %>% stringr::str_sub(end = -2)
  print(feature.name)
  minmax <- c(min(spatial.seurat[[feature]]), max(spatial.seurat[[feature]]))
  (spatial.seurat %>% 
      SpatialPlot(features = feature, crop = F, pt.size = 1.35, alpha = c(0.8, 1), images = c('sham', 'D1', 'D3', 'D7'), image.alpha = 0, stroke = 0.01) & 
      theme(text = element_text(family = plot.font, size = 9), legend.key.size = unit(0.2, 'cm')) #& NoLegend() 
    & ggplot2::scale_fill_gradientn(colours = RColorBrewer::brewer.pal(name = "RdGy", n = 9) %>% rev, limits = minmax)
    ) %>% 
  ggsave(filename = paste0(ws, "/SpatialPlot_snMG_", feature.name, ".tiff"), device = 'tiff', dpi = 300, width = 18, height = 6, units = 'cm')
})

## plot the combined spatial+vln plot
lapply(c("snMG_homeostatic1", "snMG_Reactive2"), function(features){
  pairs <- lapply(c('sham', 'D1', 'D3', 'D7'), function(time){
    subset.spatial <- spatial.seurat %>% subset(subset = Condition == time)
    minmax <-  c(spatial.seurat[[features]] %>% min, spatial.seurat[[features]] %>% max)
    p1 <- subset.spatial %>% SpatialPlot(features = features, images = time, image.alpha = 0.2, crop = F, max.cutoff = 'q99', pt.size.factor = 1.5, stroke = 0.05, alpha = c(0.5, 1)) & viridis::scale_fill_viridis(option = 'rocket', direction = -1, end = 0.95, limits = minmax) & labs(fill = features) & NoLegend() &
      theme(text = element_text(family = "OpenSans", size = 16), axis.text = element_text(family = 'OpenSans', size = 10), 
            plot.margin = unit(c(0,0,0,0), "cm"), 
            plot.background = element_blank(), 
            panel.background = element_blank())
    
    p2 <- VlnPlot(subset.spatial, features = features, pt.size = 0, group.by = "DetailedRegionAnnoShort", cols = col.list[["cols_mono_short"]], ncol = 1) & coord_flip() & scale_y_continuous(breaks = scales::breaks_pretty(n = 3)) & xlab("") & labs(title = "") & ylim(minmax) &
    theme(axis.text.x = element_text(family = 'OpenSans', angle = 30, size = 5, hjust = 0.7, vjust = 0.8), 
          axis.text.y = element_text(family = 'OpenSans', size = 10),
          text = element_text(family = "OpenSans", size = 25), 
          legend.position = "none", 
          plot.margin = unit(c(0, 0.3, 0.3, 0), "cm"),
          plot.background = element_blank(), 
          panel.background = element_blank())
    p <- grid.arrange(p1, p2, ncol = 2, widths = c(1,0.5))
  })
  feature.name <- features %>% stringr::str_sub(end = -2)
  gridExtra::grid.arrange(grobs = pairs, ncol = 4) %>% ggsave(filename = paste0(ws, "/SpatialPlot_VlnPlot_snMG_", feature.name, ".tiff"), device = 'tiff', dpi = 300, bg = NULL, width = 20, height = 4)
})
```

Comparing the reactive population with the published datasets.
```{r metanalysis}
metadata.list.full <- read_all_sheets(file = "data/SupplementaryTable4_Metanalysis_GlialReactivityMarkers.xlsx") %>% 
  .[names(.) %>% grep(pattern = "_MG_", ignore.case = T)] ## keep the microglia-related populations

metadata.list <- lapply(names(metadata.list.full), function(meta){ ##keeping genes that are in our normalized matrix
  positions <- metadata.list.full[[meta]][[1]] %in% (mg.seurat@assays$SCT %>% rownames())
  keepgenes <- metadata.list.full[[meta]][[1]][positions]
  return(keepgenes)
}); names(metadata.list) <- names(metadata.list.full)

## our sn TAM markers
ourTAMs <- read_all_sheets(file = file.path(ws, "Markers_sn_MG.xlsx"))[["SN_MG_Reactive"]] %>% 
  filter(p_val_adj < 0.01 & avg_log2FC > 0.58)
rmg.markers <- ourTAMs %>% pull(gene)

meta.intersects <- lapply(names(metadata.list), function(meta){
  paste0("Length of metadataset ", meta, " is ",  metadata.list[[meta]] %>% length, " genes.") %>% print
  int <- intersect(rmg.markers, metadata.list[[meta]]) %>% length
  paste0("Intersection is ", int) %>% print
  return(int)
  }); names(meta.intersects) <- names(metadata.list)



## signif testing
meta.signif <- lapply(names(meta.intersects), function(meta){
  subset.reac.mg <- mg.seurat %>% subset(subset = cell_type_2 == "Reactive")
  background.size <- ((subset.reac.mg@assays$RNA %>% rowSums()) > 10) %>% sum ##background gene set size with UMI counts > 10
  print(background.size)
  intersection <- meta.intersects[[meta]]
  meta.set.size <- metadata.list[[meta]] %>% length
  
  phyper(
  q = intersection - 1,  ## number of red marbles in the draw -1 (see below)
  m = meta.set.size, ## number of red marbles in urn
  n = background.size - meta.set.size, ## number of green marbles in urn
  k = rmg.markers %>% length, ## number of drawn marbles
  lower.tail = FALSE ## compute P(X > overlap), hence the '-1' in the q argument
) -> test
  return(test)
}); names(meta.signif) <- names(meta.intersects); meta.signif %<>% p.adjust(method = "fdr")
meta.signif; meta.signif < 0.01  ## see results; all below


# Reactive mg and SF's ARMs ####
## What genes/processing are different between our and ARM microglia?
library(clusterProfiler); library("org.Mm.eg.db", character.only = TRUE); organism <- org.Mm.eg.db
subset.reac.mg <- mg.seurat %>% subset(subset = cell_type_2 == "Reactive")
background <- rownames(subset.reac.mg@assays$RNA)[((subset.reac.mg@assays$RNA %>% rowSums()) > 10)] ##background gene set size with UMI counts > 10

## Unique genes+processes in our reactive MG
int.genes <- intersect(rmg.markers, metadata.list[["SF2019_MG_ARM_5XFAD"]])
rmg.markers.unique <- rmg.markers[!(rmg.markers %in% int.genes)] 
rmg.markers.unique
GO.rmg.markers.unique <- enrichGO(gene          = rmg.markers.unique, # list of genes 
                                  universe      = background, # 'potential DEGs'
                                  keyType       = "SYMBOL", # or ENSEMBL
                                  OrgDb         = organism,
                                  ont           = "ALL", # BP = biological process, MF = molecular function, CC = cellular component
                                  pAdjustMethod = "fdr",
                                  pvalueCutoff  = 0.01, 
                                  qvalueCutoff  = 0.2,
                                  readable      = F,
                                  minGSSize     = 5) %>% clusterProfiler::simplify()
View(GO.rmg.markers.unique@result)

## Unique genes+processes in SF's ARMs
arm.markers.unique <- metadata.list[["SF2019_MG_ARM_5XFAD"]][!(metadata.list[["SF2019_MG_ARM_5XFAD"]] %in% int.genes)]
arm.markers.unique
GO.arm.markers.unique <- enrichGO(gene          = arm.markers.unique, # list of genes 
                                  universe      = background, # 'potential DEGs'
                                  keyType       = "SYMBOL", # or ENSEMBL
                                  OrgDb         = organism,
                                  ont           = "ALL", # BP = biological process, MF = molecular function, CC = cellular component
                                  pAdjustMethod = "fdr",
                                  pvalueCutoff  = 0.01, 
                                  qvalueCutoff  = 0.2,
                                  readable      = F,
                                  minGSSize     = 5) %>% clusterProfiler::simplify()
View(GO.arm.markers.unique@result)


## Metadatasets in spatial ####
## add metadata
spatial.seurat %<>% AddModuleScore(features = metadata.list, name = metadata.list %>% names)
features <- names(metadata.list) %>% paste0(., seq(1:length(.))) ## feature names as they are added
identical(features, spatial.seurat[[]] %>% names %>% tail(length(features))) ## TRUE = vector names are identical

## correlation our TAMs and SF's ARM
spatial.seurat %<>% AddModuleScore(features = list(rmg.markers %>% head(15)), name = "snMG_Reac")

spatial.seurat[[]] %>% names
df <- data.frame(
  "Reactive" = spatial.seurat$snMG_Reac1,
  "SF_ARM" = spatial.seurat$SF2019_MG_ARM_5XFAD1
)

cor.test(x = df$Reactive, y = df$SF_ARM, method = "spearman")
plot(x = df$Reactive, y = df$SF_ARM)

(ggplot(data = df, aes(x = Reactive, y = SF_ARM)) + 
  geom_point(aes(color = Reactive), size = 0.5) + scale_color_gradientn(colors = c("#FC9272","#A50F15")) + xlab("Reactive Microglia signature") + ylab("ARM signature") + 
  geom_smooth(method = 'lm', se = F, color = "black") + theme_light()+ theme(axis.title = element_text(size = 20, family = "OpenSans")) + NoLegend()) %>% ggsave(filename = file.path(ws, "ScatterPlot_TAM_SFarm.tiff"), device = 'tiff', dpi = 300, height = 6, width = 6, units = "cm")


## spatialplot
features
## plot the spatial plot alone
lapply(c("SF2019_MG_ARM_5XFAD1", "SF2019_MG_IRM_5XFAD1"), function(feature){
  feature.name <- feature %>% stringr::str_sub(end = -2)
  print(feature.name)
  minmax <- c(min(spatial.seurat[[feature]]), max(spatial.seurat[[feature]]))
  (spatial.seurat %>% 
      SpatialPlot(features = feature, crop = F, pt.size = 1.35, alpha = c(0.8, 1), images = c('sham', 'D1', 'D3', 'D7'), image.alpha = 0, stroke = 0.01) & 
      theme(text = element_text(family = plot.font, size = 9), legend.key.size = unit(0.2, 'cm')) #& NoLegend() 
    & ggplot2::scale_fill_gradientn(colours = RColorBrewer::brewer.pal(name = "RdGy", n = 9) %>% rev, limits = minmax)
    ) %>% 
  ggsave(filename = paste0(ws, "/SpatialPlot_metanalysis_", feature.name, ".tiff"), device = 'tiff', dpi = 300, width = 18, height = 6, units = 'cm')
})

```

MG markers expression in UMAP
```{r Feature plot}
{# plotting microglial markers
features <- c("Csf1r", "Cd68", 'Hexb', 'Aif1', 'Lyz2', 'Spp1', 'Cst7', 'Cenpa', 'B2m', "Ctsb", 'Apoe')
lapply(features, function(feature){
  p <- (mg.seurat %>% 
      FeaturePlot(features = feature, label = T, pt.size = 1, max.cutoff = 'q99')) &
      theme(text = element_text(family = 'OpenSans')) &
      viridis::scale_color_viridis(option = 'rocket', direction = -1, alpha = 1, end = 0.9)
  p %>% ggsave(filename = paste0(ws, "/FeaturePlot_mg_marker_", feature, ".tiff"), device = 'tiff', dpi = 300, width = 8, height = 8, units = 'cm')
})
}
```

MG markers expression in the spatial plot
```{r Spatial plot}
{## spatial plot of microglial markers
  features <- c("Csf1r", "Cd68", 'Hexb', 'Aif1', 'Lyz2', 'Spp1', 'Cst7', 'Cenpa', 'B2m', 'Ctsb', 'Apoe', 'Trem2')
  lapply(features, function(feature){
    minmax <- c(spatial.seurat@assays$SCT[feature] %>% min, spatial.seurat@assays$SCT[feature] %>% max)
    p <- (spatial.seurat %>% 
      SpatialPlot(features = feature, crop = F, pt.size.factor = 1.5, image.alpha = 0.2, alpha = c(0.8, 1), stroke = 0) & 
      NoLegend() &
      theme(text = element_text(family = 'OpenSans')) & ggplot2::scale_fill_gradientn(colours = brewer.pal(n = 11, name = "RdBu") %>% rev)
    )
    p %>% ggsave(filename = paste0(ws, "/SpatialPlot_mg_marker_", feature, ".tiff"), device = 'tiff', dpi = 300, width = 20, height = 5, units = 'cm')
  })
}
```

Immunovalidation Iba1 + Ctsb
```{r immuno microglia}
immuno <- 
  read_all_sheets(
    file = file.path("results", "PNAS", "Tables", "MCAO_IHC_Validations.xlsx")
    ) %>% 
  .[["Iba1_Ctsb"]] %>% 
  mutate(Proportion_Colocalized = Proportion_Colocalized * 100) %>% 
  dplyr::mutate(Condition = factor(Condition, levels = sections)) %>% 
  group_by(SampleName, Condition) %>% 
  dplyr::summarise(
    mean_colocalization = mean(Proportion_Colocalized), 
    .groups = "drop"
  )

p.values <- 
  pairwise.wilcox.test(
    x = immuno$mean_colocalization, 
    g = immuno$Condition, 
    p.adjust.method = "BH", 
    paired = F)

immuno_summary <- 
  immuno %>% 
  group_by(Condition) %>% 
  dplyr::summarise(
    sd_colocalization = stats::sd(mean_colocalization),
    mean_colocalization = mean(mean_colocalization)
  )

# stat test control vs injured
stat_test <- 
  immuno %>% 
  mutate(Injured =  if_else(Condition == "Ctrl", "NoMCAO", "MCAO")) %>% 
  wilcox.test(mean_colocalization ~ Injured, data = .)
print(stat_test)


# plot
seurat.plots[["Barplot_Immuno_Microglia"]] <- 
  ggplot(
    data = immuno, 
    aes(y = mean_colocalization, x = Condition)
    ) + 
  geom_bar(data = immuno_summary, stat = "identity", width = 0.6, colour = "black", alpha = 0.75, fill = "#A50F15", size = 0.5) +
  geom_errorbar(
    data =  immuno_summary, 
    aes(ymin = mean_colocalization - sd_colocalization, 
        ymax = mean_colocalization + sd_colocalization), width = 0.1, color = "black", size = 0.5) +
  geom_jitter(position = position_jitter(0.2), color = 'black', size= 1.5, shape = 1, stroke = 0.5) + 
  annotate(geom = "text", x = 1, y = 20, label = paste0("P = ", stat_test$p.value %>% round(3)), vjust = -0.5, hjust = 0, size = 2.45) +
  ylim(0, 35) +
  xlab(NULL) + 
  ylab("IBA1 + CSTB co-localization (%)") +
  theme_mk + 
  remove_grid + 
  theme(panel.grid.major.y = element_line(linewidth = 0.1, colour = "grey"))

print(seurat.plots[["Barplot_Immuno_Microglia"]])

png_save_show(
  plot = seurat.plots[["Barplot_Immuno_Microglia"]], 
  file = file.path(ws, "BarPlot_microglia_Immuno.png"), 
  show_plot = FALSE, 
  dpi = 1000, 
  height = 70, 
  width = 50
)

```


=== Oligodendrocytes ===
## single-nucleus OLs
 
```{r ol results and plot lists}
oligo.results <- list()
oligo.plots <- list()
```

```{r subset and processing sn OLs}
sn.seurat %<>% SetIdent(value = "Celltypes")
oligo.seurat <- subset(sn.seurat, idents = c("OLs", "OPCs"))

# Normalize and scale
{
  DefaultAssay(oligo.seurat) <- "RNA"
  oligo.seurat %<>% NormalizeData(normalization.method = "LogNormalize", verbose = T)
  oligo.seurat %<>% FindVariableFeatures(selection.method = "vst", nfeatures = 3000)
  oligo.seurat %<>% ScaleData()
}

# Explore gene distribution across cells
oligo.seurat[["RNA"]]@counts %>% rowSums() %>% summary
oligo.seurat[["RNA"]]@counts %>% rowSums() %>% sort(decreasing = T) %>% head(50) ## the most expressed genes overall

# Gene presence and variability in cells
oligo.results[["GenesAboveFive"]] <- (oligo.seurat[["RNA"]]@counts %>% rowSums()) %>% .[. >= 5 ] ## no genes in >= 5 cells
oligo.results[["rna_varfeat"]] <- oligo.seurat %>% HVFInfo(assay = "RNA")
oligo.results[["rna_varfeat"]] %>% arrange(desc(variance.standardized)) %>% top_n(20)
selected.var.features <- oligo.results[["rna_varfeat"]] %>% dplyr::filter(variance.standardized >= 1.5) %>% rownames ## highly variable genes
((selected.var.features %in% names(oligo.results[["GenesAboveFive"]])) %>% sum) / length(selected.var.features) ## >99% of var genes are present in at least five cells

oligo.results[["GenesAboveFive"]] %>% .[names(.) %in% selected.var.features] %>% hist(breaks = 1000)

# PCA
{
  oligo.seurat %<>% RunPCA(assay = "RNA", features = selected.var.features, npcs = 50, verbose = T)
  oligo.seurat %>% DimPlot(reduction = "pca", dims = c(1,2), cols = col.list$Celltypes)
  oligo.seurat %>% ElbowPlot(ndims = 50) | oligo.seurat %>% PC_var_explained()
  oligo.seurat %>% DimHeatmap(cells = 200, reduction = "pca", balanced = T, dims = 1:15)
}

# the 3D PCA
{
  dims <- c(1, 2, 3) ## PCs to plot
  df <- data.frame(
      "ElementName" = oligo.seurat@reductions$pca@cell.embeddings %>% rownames(),
      "X" = oligo.seurat@reductions$pca@cell.embeddings[, dims[1]],
      "Y" = oligo.seurat@reductions$pca@cell.embeddings[, dims[2]],
      "Z" = oligo.seurat@reductions$pca@cell.embeddings[, dims[3]],
      "Annotation" = oligo.seurat[["CelltypesDetailed"]]
  )
  df %<>% sample_n(size = 1000) # (optional) to increase plotting speed
  df %>% plot_ly(x = .[["X"]], y = .[["Y"]], z = .[["Z"]], type="scatter3d", mode = "markers", color = .[["Annotation"]], colors = col.list$CelltypesDetailed)
}

## UMAP
pc_dims <- 1:10
{
  oligo.seurat %<>% RunUMAP(reduction = "pca", dims = pc_dims, verbose = T)
  oligo.seurat %>% DimPlot(reduction = "umap", group.by = "CelltypesDetailed", label = T, cols = col.list$CelltypesDetailed)
}

## Clustering 
{
  oligo.seurat %<>% FindNeighbors(reduction = "pca", dims = pc_dims, verbose = T)
  for(res in c(0.6, 0.8, 1, 1.2, 1.4)){
   oligo.seurat %<>% FindClusters(resolution = res, algorithm = 1, verbose = T) 
  }
}

oligo.plots[["UMAP_RNA_res1.4"]] <- oligo.seurat %>% 
  DimPlot(reduction = "umap", pt.size = 1, group.by = "RNA_snn_res.1.4", cols = col.list$big_col_palette, label = T)

oligo.seurat %>% FeaturePlot(features = "Serpina3n", pt.size = 1.5)

oligo.seurat %>% dittoPlot(var = "Tnfrsf12a", group.by = "RNA_snn_res.1.4", plots = c("vlnplot", "boxplot", "jitter"))
```
 
```{r markers}
DefaultAssay(oligo.seurat) <- "RNA"
oligo.seurat %<>% SetIdent(value = "RNA_snn_res.1.4")

oligo.results[["markers_RNAres1.4"]] <- FindAllMarkers(oligo.seurat, logfc.threshold = 1, assay = 'RNA', verbose = T, only.pos = T) %>% 
  arrange(p_val_adj) %>% 
  filter(p_val_adj < 0.01) %>% 
  split(f = .[["cluster"]])

```

```{r Annotating sn OLs}
canonical.features <- c(
  "Plp1", "Mobp", ## OL-lineage cells
  "Nckap5", "Tcf7l2", ## NFOL
  "Sepp1", "S100b", ## MOL2
  "Opalin", "Ptgds", ## MOL5/6
  "Ifi27", "H2-D1", ## IFN-responsive
  "Serpina3n", "Klk6", ## MOL DA1
  "Tnfrsf12a", "Cdkn1a" ## MOL DA2
)

oligo.plots[["FeaturePlots_sn_CanonicalMarkers"]] <- oligo.seurat %>% 
  FeaturePlot(features = canonical.features, max.cutoff = 'q99', repel = T, label.size = 3, slot = "data",
  reduction = "umap", label = T, raster = F, pt.size = 0.2, ncol = 5, cols = c("white", "#AE0900")) & NoAxes() & NoLegend()


oligo.seurat@meta.data %<>% mutate("OLCelltypes" = factor(case_when(
  RNA_snn_res.1.4 == 0 ~ "MOL5/6", 
  RNA_snn_res.1.4 == 1 ~ "MOL5/6", 
  RNA_snn_res.1.4 == 2 ~ "MOL2", 
  RNA_snn_res.1.4 == 3 ~ "MOL2", 
  RNA_snn_res.1.4 == 4 ~ "OPCs",
  RNA_snn_res.1.4 == 5 ~ "MOL5/6", 
  RNA_snn_res.1.4 == 6 ~ "MOL5/6", 
  RNA_snn_res.1.4 == 7 ~ "NFOLs", 
  RNA_snn_res.1.4 == 8 ~ "MOL5/6", 
  RNA_snn_res.1.4 == 9 ~ "OPCs", 
  RNA_snn_res.1.4 == 10 ~ "MOL_DA1", 
  RNA_snn_res.1.4 == 11 ~ "MOL5/6", 
  RNA_snn_res.1.4 == 12 ~ "OPCs"
), levels = c("OPCs", "NFOLs", "MOL2", "MOL5/6", "MOL_DA1")))

oligo.seurat %<>% SetIdent(value = "OLCelltypes")
```

```{r Plotting sn OLs}
oligo.results[["UMAP_snOLs_RNAres1.4"]] <- oligo.seurat %>% DimPlot(group.by = "RNA_snn_res.1.4", pt.size = 1)

oligo.plots[["UMAP_snOLs_OLCelltypes"]] <- oligo.seurat %>% 
  DimPlot(group.by = "OLCelltypes", pt.size = 1.5, cols = col.list$ol_cols, label = T, label.box = 5,  label.size = 5) & NoAxes() & theme(legend.position = "none")

oligo.plots[["Barplot_snOLs_Conditions_byOLCelltypes"]] <- oligo.seurat %>% 
  dittoBarPlot(var = "Condition", group.by = "OLCelltypes", color.panel = col.list[["Condition"]], retain.factor.levels = T) 
```

```{r save sn OLs}
saveRDS(oligo.seurat, file = "data/Seurat_snOLs_ReadyForIntegration.Rds")
```

###
Load the seurat
```{r seurat load}
if(!dir.exists("results/SpatialFigures/Figure4/oligodendrocytes")){dir.create("results/SpatialFigures/Figure4/oligodendrocytes"); ws <- "results/SpatialFigures/Figure4/oligodendrocytes"}else{ws <- "results/SpatialFigures/Figure4/oligodendrocytes"}

oligo.seurat <- readRDS("data/seurat_OLIGO_names_soupX.rds"); DefaultAssay(oligo.seurat) <- "SCT"
oligo.seurat$cell_type_2 %>% levels()
oligo.seurat %<>% subset(idents = "dead", invert = T); oligo.seurat$cell_type_2 %<>% droplevels() ## remove dead cells and their factor
oligo.seurat$Condition %<>% factor(levels = c("sham", "D1", "D3", "D7"))

oligo.seurat[["cell_type_2"]] <- case_when(
  oligo.seurat$cell_type_2 == "OPC" ~ "OPC",
  oligo.seurat$cell_type_2 == "COP+NFOL" ~ "NFOL",
  oligo.seurat$cell_type_2 == "MFOL+MOL" ~ "MOL",
  oligo.seurat$cell_type_2 == "MOL" ~ "MOL",
  oligo.seurat$cell_type_2 == "FastResponseOL" ~ "MOL",
  oligo.seurat$cell_type_2 == "TAO" ~ "Reactive",
) %>% factor(levels = c("OPC", "NFOL", "MOL", "Reactive")); oligo.seurat %<>% SetIdent(value = "cell_type_2")

oligo.seurat[[]] %>% names

col.list[["ol_cols"]] <- c(
  "OPC" = "#BFD3E6",
  "NFOL" = "#9EBCDA",
  "MOL" = "#8C96C6",
  "Reactive" = "#810F7C"
)
```

UMAP
```{r Figure sn UMAP}
(oligo.seurat %>% 
  UMAPPlot(group.by = "cell_type_2", pt.size = 1, label = F, label.box = T, cols = col.list[["ol_cols"]]) + 
  NoLegend() + 
  theme(plot.background = element_blank(), panel.background = element_blank())) %>% ggsave(filename = file.path(ws, "UMAPPlot_sn_ol.tiff"), device = 'tiff', dpi = 300, units = 'cm', width = 10, height = 10)
```

Proportions per condition
```{r Proportions}
## proportions
library(dittoSeq)
oligo.seurat$Condition %>% table
oligo.seurat$cell_type_2 %>% table

oligo.seurat %>% dittoBarPlot(var = "Condition", group.by = "cell_type_2", scale = "count")
oligo.seurat %>% dittoBarPlot(var = "Condition", group.by = "cell_type_2", scale = "percent")

## data frames
ids <- oligo.seurat@meta.data$Condition %>% levels; categories <- levels(oligo.seurat@meta.data$cell_type_2); counts <- matrix(nrow=length(ids), ncol=length(categories)); rownames(counts) <- ids; colnames(counts) <- categories
for (i in seq_along(ids)) {
  for (j in seq_along(categories)) {
    count <- oligo.seurat@meta.data %>%
      filter(Condition == ids[i], cell_type_2 == categories[j]) %>%
      nrow()

    counts[i, j] <- count
  }
}; rm(i,j)

counts_mm_0 <- reshape2::melt(counts, id = "rownames") ## total count proportions

ggplot(data = counts_mm_0, aes(x = Var2, y = value, fill = Var1)) + 
  geom_col(position = "dodge") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  ylab('Proportion of cells')
counts <- counts/rowSums(counts)
counts_mm_1 <- reshape2::melt(counts, id = "rownames") ## percent count proportions


p <-  ggplot(data = counts_mm_1, aes(x = Var2, y = value, fill = Var1)) + ## percent count proportions
    geom_col(position = "dodge") +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 105, hjust = 1, vjust = 0, family = "OpenSans", size = 35, color = "black"), 
          axis.text.y = element_text(angle = 90, hjust = 0.5, vjust = 0.5, family = "OpenSans", size = 18, color = "black"),
          axis.title = element_text(family = "OpenSans", size = 25),
          plot.background = element_blank(),
          panel.background = element_blank()) +
    ylab('Proportion of nuclei') +
    xlab("") + ggtitle("") + NoLegend() + ylim(0,1) +
    scale_fill_manual(values = c("#525475", "#F1D7AB", "#E6B188", "#D18080")); p
p %>% ggsave(filename = file.path(ws, "BarPlot_PercentProportions_sn_OL_bySample.tiff"), device = 'tiff', units = "cm", dpi = 300, width = 6, height = 8, bg = 'transparent')

```

Markers, dotplot
```{r Markers, Dotplot}
## Markers
ol.seurat.markers <- FindAllMarkers(object = oligo.seurat, assay = "SCT", logfc.threshold = 0.58, verbose = T, only.pos = T) %>% 
  arrange(desc(avg_log2FC)) %>% 
  filter(p_val_adj < 0.01) %>% 
  split(f = .[["cluster"]])

##adding a column on marker's median expression for the population
ol.markers2 <- lapply(names(ol.seurat.markers), function(pop){
  temppos <- oligo.seurat$cell_type_2 == pop
  temp <- oligo.seurat@assays$RNA@counts[, temppos] %>% apply(MARGIN = 1, FUN = median)
  tempgenes <- ol.seurat.markers[[pop]][["gene"]]
  temp %<>% .[names(temp) %in% tempgenes] %>% .[match(tempgenes, table = names(.))]
  ol.seurat.markers[[pop]] %<>% dplyr::mutate("MedianPopulationUMI" = temp)
  return(ol.seurat.markers[[pop]])
}); names(ol.markers2) <- names(ol.seurat.markers)

ol.markers2 %>% write.xlsx(file = file.path(ws, "Markers_sn_OL.xlsx"), overwrite = T)

## markers of reactive vs MOL
pops <- c("Reactive")
reac.ol.list <- lapply(pops, function(OL){
  temp <- FindMarkers(oligo.seurat, ident.1 = OL, ident.2 = "MOL", assay = "SCT", logfc.threshold = 0.58, only.pos = T, verbose = T) %>% 
  arrange(desc(avg_log2FC)) %>% 
  filter(p_val_adj < 0.01) %>% 
  rownames_to_column(var = "gene")
  return(temp)
}); names(reac.ol.list) <- pops


## dotplot
{
features_test <- c("Pdgfra", "Sox6", "Cspg4",  #OPC
              "Rras2", "Prom1", "Tcf7l2", "Nckap5", #NFOL
              "Plp1", "Mog", "Mast4",  "Grm3", "Ptgds", #MOL
              "Pkp4", "Neat1", "Rhoj", "Il33", "C4b", "Gab2", "Serpina3n"  #Reactive
              )

oligo.seurat %>% FeaturePlot(features = features_test, max.cutoff = 'q99')
oligo.seurat %>% VlnPlot(features = features_test)

p <- DotPlot(oligo.seurat, assay = "SCT", features = features_test, group.by = "cell_type_2", cols = c("#FFF7FB", "#8C96C6"), scale = F) + 
  theme(text = element_text(family = "OpenSans", size = 20), 
        axis.text.y = element_text(size = 30, hjust = 1), 
        axis.text.x = element_text(angle = 55, vjust = 1, hjust = 1, size = 25), legend.position = "right",
        axis.ticks = element_blank(), 
        axis.title = element_blank());p
p %>% ggsave(filename = paste0(ws, "/DotPlot_sn_OL.tiff"), device = 'tiff', dpi = 300, width = 5, height = 3)  
}

```

Gene ontology analysis
```{r GO analysis}
ol.markers <- read_all_sheets(file.path(ws, "Markers_sn_OL.xlsx"))

reac.degs <- oligo.seurat %>% FindMarkers(ident.1 = "Reactive", ident.2 = "MOL", logfc.threshold = 0.58, assay = "SCT", verbose = T, only.pos = T)
reac.markers <- list(
  "snReactive" = reac.degs %>% filter(avg_log2FC > 0.58) %>% rownames() %>% .[grep("^Rp|^mt", x = ., ignore.case = F, invert = T)]
)


## GO analysis
library(clusterProfiler); library("org.Mm.eg.db", character.only = TRUE)
organism <- org.Mm.eg.db; background <- rownames(oligo.seurat[["SCT"]])
ORA.list <- lapply(reac.markers, function(x) enrichGO(gene          = x, # list of genes 
                                                      universe      = background, # 'potential DEGs'
                                                      keyType       = "SYMBOL", # or ENSEMBL
                                                      OrgDb         = organism,
                                                      ont           = "ALL", # BP = biological process, MF = molecular function, CC = cellular component
                                                      pAdjustMethod = "fdr",
                                                      pvalueCutoff  = 0.01, #
                                                      qvalueCutoff  = 0.2,
                                                      readable      = F,
                                                      minGSSize     = 3) #%>% clusterProfiler::simplify()
                   )
ORA.list %>% write.xlsx(file = file.path(ws, "ORAlist_sn_ReacOligo.xlsx"), overwrite = T)

## selecting terms for the barplot
terms <- c(
           ## BP
           "", # 

           ## CC
           "", # 

           ## MF
           "" # 
           )
```

Sn signature in spatial
```{r sn signature in spatial}
reac.degs <- oligo.seurat %>% FindMarkers(ident.1 = "Reactive", ident.2 = "MOL", logfc.threshold = 0.58, assay = "SCT", verbose = T, only.pos = T)
reac.markers <- list(
  "snReactive" = reac.degs %>% filter(avg_log2FC > 0.58) %>% rownames() %>% .[grep("^Rp|^mt", x = ., ignore.case = F, invert = T)] %>% head(15)
)

spatial.seurat %<>% AddModuleScore(features = reac.markers, name = names(reac.markers))
spatial.seurat[[]] %>% names

features <- c("snReactive1")
## plot the combined spatial+vln plot
lapply(features, function(feature){
  feature.name <- feature %>% stringr::str_sub(end = -2)
  print(feature.name)
  minmax <- c(min(spatial.seurat[[feature]]), max(spatial.seurat[[feature]]))
  (spatial.seurat %>% 
      SpatialPlot(features = feature, crop = F, pt.size = 1.35, alpha = c(0.8, 1), images = c('sham', 'D1', 'D3', 'D7'), image.alpha = 0, stroke = 0.01) & 
      theme(text = element_text(family = plot.font, size = 9), legend.key.size = unit(0.2, 'cm')) & NoLegend() 
    & ggplot2::scale_fill_gradientn(colours = c("#4D4D4D", "#FFFFFF", "#211E61"), limits = minmax)
    ) %>% 
  ggsave(filename = paste0(ws, "/SpatialPlot_scsnOL_", feature.name, ".tiff"), device = 'tiff', dpi = 300, width = 18, height = 6, units = 'cm')
})
```

Published OL datasets in the single nucleus and spatial data.
```{r metanalysis}
##snOL markers
reac.degs <- oligo.seurat %>% FindMarkers(ident.1 = "Reactive", ident.2 = "MOL", logfc.threshold = 0.58, assay = "SCT", verbose = T, only.pos = T) %>% arrange(desc(avg_log2FC))
reac.markers <- list(
  "snReactive" = reac.degs %>% filter(avg_log2FC > 0.58) %>% rownames() %>% .[grep("^Rp|^mt", x = ., ignore.case = F, invert = T)]
)
reac.degs[["avg_log2FC"]] %>% min ## only positive markers
rol.markers <- reac.markers[["snReactive"]]

## metadata
metadata.list <- read_all_sheets(file = "data/SupplementaryTable4_Metanalysis_GlialReactivityMarkers.xlsx") %>% 
  .[names(.) %>% grep(pattern = "MOL|_OL_|DOL", ignore.case = T)] %>%  ## keep the OL-related populations
  lapply(., function(df) df %>% pull()) ## change the dataframes into vectors

oligo.seurat %<>% AddModuleScore(features = metadata.list, name = names(metadata.list))
oligo.seurat[[]] %>% names

features <- oligo.seurat[[]] %>% names %>% tail(length(metadata.list))


## overlap testing
meta.intersects <- lapply(names(metadata.list), function(meta){
  paste0("Length of metadataset ", meta, " is ",  metadata.list[[meta]] %>% length, " genes.") %>% print
  int <- intersect(rol.markers, metadata.list[[meta]]) %>% length
  paste0("Intersection is ", int) %>% print
  return(int)
  }); names(meta.intersects) <- names(metadata.list)

## signif testing gene overlaps
meta.signif <- lapply(names(meta.intersects), function(meta){
  background.size <- ((oligo.seurat@assays$RNA %>% rowSums()) > 10) %>% sum ##background gene set size with UMI counts > 10
  print(background.size)
  intersection <- meta.intersects[[meta]]
  meta.set.size <- metadata.list[[meta]] %>% length
  
  phyper(
  q = intersection - 1,  ## number of red marbles in the draw -1 (see below)
  m = meta.set.size, ## number of red marbles in urn
  n = background.size - meta.set.size, ## number of green marbles in urn
  k = rol.markers %>% length, ## number of drawn marbles
  lower.tail = FALSE ## compute P(X > overlap), hence the '-1' in the q argument
) -> test
  return(test)
}); names(meta.signif) <- names(meta.intersects); meta.signif %<>% p.adjust(method = "fdr")
meta.signif; meta.signif < 0.01  ## see results; all below


## Published datasets in spatial
## add metadata
spatial.seurat %<>% AddModuleScore(features = metadata.list, name = metadata.list %>% names)
features <- names(metadata.list) %>% paste0(., seq(1:length(.))) ## feature names as they are added
identical(features, spatial.seurat[[]] %>% names %>% tail(length(features))) ## TRUE = vector names are identical

## correlation Reactive and DOL
spatial.seurat %<>% AddModuleScore(features = list(rol.markers %>% head(15)), name = "snOL_Reac")
spatial.seurat[[]] %>% names

df <- data.frame(
  "snOL_Reac" = spatial.seurat$snOL_Reac1,
  "Kenigsbuch_DOL" = spatial.seurat$Kenigsbuch2022_AD_DOL6
)

cor.test(x = df$snOL_Reac, y = df$Kenigsbuch_DOL, method = "spearman")
plot(x = df$snOL_Reac, y = df$Kenigsbuch_DOL)

(ggplot(data = df, aes(x = snOL_Reac, y = Kenigsbuch_DOL)) + 
  geom_point(aes(color = snOL_Reac), size = 0.5) + scale_color_gradientn(colors = c("#8C96C6","#810F7C")) + xlab("Reactive Oligodendrocytes signature") + ylab("DOL signature") +
  geom_smooth(method = 'lm', se = F, color = "black") + theme_light()+ theme(axis.title = element_text(size = 20, family = "OpenSans")) + NoLegend()) %>% ggsave(filename = file.path(ws, "ScatterPlot_snOLReac_KenigsbuchDOL.tiff"), device = 'tiff', dpi = 300, height = 6, width = 7, units = "cm")

```



## single-cell OLs

Subset and processing sc OLs
```{r subset and processing sc OLs}
oligo.integrated <- readRDS("data/seurat_OLIGO_integrated_UMAP_named.rds")
oligo.integrated %<>% subset(subset = orig.ident == "scRNA_MCAO_all")

ol.assay <- oligo.integrated %>% GetAssayData(assay = "RNA", layer = "counts")
ol.metadata <- oligo.integrated@meta.data[c("orig.ident", "Sample", "Condition", "Doublet", "cell_type_1","Phase", "percent.mt", "percent.rib", "scsnOLAnno")]

oligo.seurat.sc <- CreateSeuratObject(counts = ol.assay, meta.data = ol.metadata, project = "Zucha2023_scOLs")
oligo.seurat.sc %<>% subset(subset = Doublet == "Singlet")
oligo.seurat.sc %<>% SetIdent(value = "scsnOLAnno")

# Normalize and scale
{
  DefaultAssay(oligo.seurat.sc) <- "RNA"
  oligo.seurat.sc %<>% NormalizeData(normalization.method = "LogNormalize", verbose = T)
  oligo.seurat.sc %<>% FindVariableFeatures(selection.method = "vst", nfeatures = 3000)
  oligo.seurat.sc %<>% ScaleData()
}

# Explore gene distribution across cells
oligo.seurat.sc[["RNA"]]@counts %>% rowSums() %>% summary
oligo.seurat.sc[["RNA"]]@counts %>% rowSums() %>% sort(decreasing = T) %>% head(50) ## the most expressed genes overall

# Gene presence and variability in cells
oligo.results[["GenesAboveFive"]] <- (oligo.seurat.sc[["RNA"]]@counts %>% rowSums()) %>% .[. >= 5 ] ## no genes in >= 5 cells
oligo.results[["rna_varfeat"]] <- oligo.seurat.sc %>% HVFInfo(assay = "RNA")
oligo.results[["rna_varfeat"]] %>% arrange(desc(variance.standardized)) %>% top_n(20)
selected.var.features <- oligo.results[["rna_varfeat"]] %>% dplyr::filter(variance.standardized >= 1.5) %>% rownames ## highly variable genes
((selected.var.features %in% names(oligo.results[["GenesAboveFive"]])) %>% sum) / length(selected.var.features) ## >99% of var genes are present in at least five cells

oligo.results[["GenesAboveFive"]] %>% .[names(.) %in% selected.var.features] %>% hist(breaks = 1000)

# PCA
{
  oligo.seurat.sc %<>% RunPCA(assay = "RNA", features = selected.var.features, npcs = 50, verbose = T)
  oligo.seurat.sc %>% DimPlot(reduction = "pca", dims = c(1,2), cols = col.list$big_col_palette)
  oligo.seurat.sc %>% ElbowPlot(ndims = 50) | oligo.seurat.sc %>% PC_var_explained()
  oligo.seurat.sc %>% DimHeatmap(cells = 200, reduction = "pca", balanced = T, dims = 1:15)
}

# the 3D PCA
{
  dims <- c(1, 2, 3) ## PCs to plot
  df <- data.frame(
      "ElementName" = oligo.seurat.sc@reductions$pca@cell.embeddings %>% rownames(),
      "X" = oligo.seurat.sc@reductions$pca@cell.embeddings[, dims[1]],
      "Y" = oligo.seurat.sc@reductions$pca@cell.embeddings[, dims[2]],
      "Z" = oligo.seurat.sc@reductions$pca@cell.embeddings[, dims[3]],
      "Annotation" = oligo.seurat.sc[["scsnOLAnno"]]
  )
  df %<>% sample_n(size = 1000) # (optional) to increase plotting speed
  df %>% plot_ly(x = .[["X"]], y = .[["Y"]], z = .[["Z"]], type="scatter3d", mode = "markers", color = .[["Annotation"]], colors = col.list$ol_cols)
}


## UMAP
pc_dims <- 1:10
{
  oligo.seurat.sc %<>% RunUMAP(reduction = "pca", dims = pc_dims, verbose = T)
  oligo.seurat.sc %>% DimPlot(reduction = "umap", group.by = "scsnOLAnno", label = T, cols = col.list$ol_cols)
}

## Clustering 
{
  oligo.seurat.sc %<>% FindNeighbors(reduction = "pca", dims = pc_dims, verbose = T)
  for(res in c(0.6, 0.8, 1, 1.2, 1.4)){
   oligo.seurat.sc %<>% FindClusters(resolution = res, algorithm = 1, verbose = T) 
  }
}

oligo.seurat.sc %<>% SetIdent(value = "RNA_snn_res.1.4")

oligo.plots[["UMAP_sc_RNA_res1.4"]] <- oligo.seurat.sc %>% 
  DimPlot(reduction = "umap", pt.size = 1, group.by = "RNA_snn_res.1.4", cols = col.list$big_col_palette, label = T)

```

```{r Annotating sc OLs}
canonical.features <- c(
  "Plp1", "Mobp", ## OL-lineage cells
  "Nckap5", "Tcf7l2", ## NFOL
  "Sepp1", "S100b", ## MOL2
  "Opalin", "Ptgds", ## MOL5/6
  "Ifi27", "H2-D1", ## IFN-responsive
  "Serpina3n", "Klk6", ## MOL DA1
  "Tnfrsf12a", "Cdkn1a" ## MOL DA2
)

oligo.plots[["FeaturePlots_sc_CanonicalMarkers"]] <- oligo.seurat.sc %>% 
  FeaturePlot(features = canonical.features, max.cutoff = 'q99', repel = T, label.size = 3, slot = "data",
  reduction = "umap", label = T, raster = F, pt.size = 0.2, ncol = 5, cols = c("white", "#AE0900")) & NoAxes() & NoLegend()


oligo.seurat.sc@meta.data %<>% mutate("OLCelltypes" = factor(case_when(
  RNA_snn_res.1.4 == 0 ~ "MOL5/6", 
  RNA_snn_res.1.4 == 1 ~ "MOL5/6", 
  RNA_snn_res.1.4 == 2 ~ "MOL5/6", 
  RNA_snn_res.1.4 == 3 ~ "MOL5/6", 
  RNA_snn_res.1.4 == 4 ~ "MOL2",
  RNA_snn_res.1.4 == 5 ~ "MOL5/6", 
  RNA_snn_res.1.4 == 6 ~ "MOL5/6", 
  RNA_snn_res.1.4 == 7 ~ "MOL5/6", 
  RNA_snn_res.1.4 == 8 ~ "MOL5/6", 
  RNA_snn_res.1.4 == 9 ~ "MOL2", 
  RNA_snn_res.1.4 == 10 ~ "MOL_DA1", 
  RNA_snn_res.1.4 == 11 ~ "MOL5/6", 
  RNA_snn_res.1.4 == 12 ~ "MOL5/6",
  RNA_snn_res.1.4 == 13 ~ "MOL_DA2",
  RNA_snn_res.1.4 == 14 ~ "MOL2"
), levels = c("MOL2", "MOL5/6", "MOL_DA1", "MOL_DA2")))


oligo.seurat.sc %<>% SetIdent(value = "OLCelltypes")
```

```{r Plotting sc OLs}
oligo.results[["UMAP_scOLs_RNAres1.4"]] <- oligo.seurat.sc %>% DimPlot(group.by = "RNA_snn_res.1.4", pt.size = 1)

oligo.plots[["UMAP_scOLs_OLCelltypes"]] <- oligo.seurat.sc %>% 
  DimPlot(group.by = "OLCelltypes", pt.size = 1.5, cols = col.list$ol_cols, label = T, label.box = 5,  label.size = 5) & NoAxes() & theme(legend.position = "none")

oligo.plots[["Barplot_scOLs_Conditions_byOLCelltypes"]] <- oligo.seurat.sc %>% 
  dittoBarPlot(var = "Condition", group.by = "OLCelltypes", color.panel = col.list[["Condition"]], retain.factor.levels = T) 
```

```{r save sc OLs}
saveRDS(oligo.seurat.sc, file = "data/Seurat_scOLs_ReadyForIntegration.Rds")
```

## Integration of sn and sc Oligos
```{r load respective ol seurats v4}
oligo.seurat <- readRDS("data/Seurat_snOLs_ReadyForIntegration.Rds")
oligo.seurat$OLCelltypes %>% table
oligo.seurat %<>% subset(subset = OLCelltypes != "OPCs")

oligo.seurat.sc <- readRDS("data/Seurat_scOLs_ReadyForIntegration.Rds")
oligo.seurat.sc$OLCelltypes %>% table

options(Seurat.object.assay.version = "v5")

seurat.list <- list(
  "sn" = oligo.seurat,
  "sc" = oligo.seurat.sc
)
```

The seurats were prepared as in Seurat v4 but for smoother integration and dataset longevity we transform them to v5.
```{r transform seurats v4->v5}
## clean up and convert to v5
seuratv5.list <- lapply(seurat.list, function(x){
  DefaultAssay(x) <- "RNA"
  x %<>% DietSeurat(assays = "RNA", layers = "counts")
  assay5 <- as(x[["RNA"]], Class = "Assay5")
  seurat5 <- CreateSeuratObject(assay5, meta.data = x@meta.data)
  return(seurat5)
})

##merge into a single big seurat v5 object
names(seuratv5.list)
merged.seurat <- merge(seuratv5.list[["sn"]], 
                      seuratv5.list[["sc"]])

## keep non-NA metadata and merge the count matrix
merged.seurat@meta.data %<>% .[ , colSums(is.na(.)) == 0]
merged.seurat %<>% JoinLayers()
Idents(merged.seurat) <- "orig.ident"

## non-near-zero genes that are in >= 5 cells with >= 5 counts
genes.to.keep <- base::intersect(rownames(merged.seurat[["RNA"]]$data)[(merged.seurat[["RNA"]]$data %>% rowSums()) >= 5], ## genes with >= 5 counts
                                 rownames(merged.seurat[["RNA"]]$data)[((merged.seurat[["RNA"]]$data > 0) %>% rowSums()) >= 5] ## genes present in >= 5 cells
                                 )

## split based on the sequencing batches
merged.seurat[["RNA"]] <- split(merged.seurat[["RNA"]], f = merged.seurat$orig.ident)
```

```{r Process without integration}
# Normalize, Scale, FindVariableFeatures
{
  gc(full = T)
  merged.seurat %<>% NormalizeData()
  merged.seurat %<>% ScaleData(features = genes.to.keep)
  merged.seurat %<>% FindVariableFeatures(nfeatures = 3000, verbose = T)
}

# PCA
{
  merged.seurat %<>% RunPCA(npcs = 50, verbose = T)
  merged.seurat %>% PCAPlot(group.by = "orig.ident", cols = col.list$big_col_palette, raster = F, shuffle = T) + NoLegend()
}

# PCA QC
PC_var_explained(Seurat = merged.seurat) | merged.seurat %>% ElbowPlot(ndims = 50)
merged.seurat %>% DimHeatmap(dims = 1:15, cells = 300, balanced = TRUE)
merged.seurat %>% DimHeatmap(dims = 16:30, cells = 300, balanced = TRUE)

# UMAP
pca.dims <- 1:15
{
  merged.seurat %<>% RunUMAP(dims = pca.dims, reduction = "pca",  verbose = T, reduction.name = "umap_unintegrated")
  merged.seurat %>% DimPlot(reduction = "umap_unintegrated", group.by = 'orig.ident', cols = col.list$big_col_palette, raster = F, combine = T, shuffle = T) ## strong batch effect is present
}
```

Strong technical batch-effect is present. Let's integrate
```{r Integrate}
gc(full = T)

merged.seurat %<>% IntegrateLayers(method = CCAIntegration,
  orig.reduction = "pca", new.reduction = "cca",
  verbose = T)

merged.seurat %<>% JoinLayers()
```

Clusters and UMAP in the integrated OL dataset
```{r integrated clusters and umap}
pca.dims <- 1:15
## cca
{
  merged.seurat %<>% FindNeighbors(reduction = "cca", dims = pca.dims) 
  merged.seurat %<>% FindClusters(resolution = 1.4, cluster.name = "cca_clusters")
  merged.seurat %<>% RunUMAP(reduction = "cca", dims = pca.dims, reduction.name = "umap.cca")
}


oligo.plots[["UMAP_int_ccaclusters"]] <-  merged.seurat %>% 
  DimPlot(reduction = "umap.cca", pt.size = 1.5, group.by = "cca_clusters", cols = col.list$big_col_palette, shuffle = T, label = T, label.box = T, label.size = 4) & NoAxes()

oligo.plots[["UMAP_int_OLCelltypes"]] <- merged.seurat %>% 
  DimPlot(reduction = "umap.cca", pt.size = 1, group.by = "OLCelltypes", cols = col.list$ol_cols, label = T, label.box = T, label.size = 4, shuffle = T)
merged.seurat %>% DimPlot(reduction = "umap.cca", pt.size = 1, group.by = "orig.ident")

merged.seurat %>% VlnPlot(features = c("Irf7", "H2-D1", "Ifi27"), cols = col.list$big_col_palette)
merged.seurat %>% VlnPlot(features = c("Serpina3n", "Tnfrsf12a"), cols = col.list$big_col_palette)
```

CCA integration maintains description observed in the two individual datasets.
```{r Canonical Markers in the integrated dataset}
canonical.features <- c(
  "Plp1", "Mobp", ## OL-lineage cells
  "Nckap5", "Tcf7l2", ## NFOL
  "Sepp1", "S100b", ## MOL2
  "Opalin", "Ptgds", ## MOL5/6
  "Irf7", "Ifi27", "H2-D1", ## IFN-responsive
  "Serpina3n", "Klk6", ## MOL DA1
  "Tnfrsf12a", "Cdkn1a" ## MOL DA2
)

oligo.plots[["FeaturePlots_int_CanonicalMarkers_cca"]] <- merged.seurat %>% 
  FeaturePlot(features = canonical.features, max.cutoff = 'q99', repel = T, label.size = 3, slot = "data",
  reduction = "umap.cca", label = T, raster = F, pt.size = 0.2, ncol = 5, cols = c("white", "#AE0900")) & NoAxes() & NoLegend()
```

Calculating markers in cca clusters
```{r OL Markers in cca clusters}
DefaultAssay(merged.seurat) <- "RNA"
merged.seurat %<>% SetIdent(value = "cca_clusters")


oligo.results[["FindAllMarkers_int_ccaclusters"]] <- FindAllMarkers(merged.seurat, logfc.threshold = 1, assay = 'RNA', verbose = T, only.pos = T) %>% 
  arrange(p_val_adj) %>% 
  filter(p_val_adj < 0.01) %>% 
  split(f = .[["cluster"]])
```

Annotating the integrated oligo dataset
```{r annotating integrated OL dataset}
merged.seurat@meta.data %<>% mutate("intOLSubtypes" = factor(case_when(
  cca_clusters == 0 ~ "MOL5/6",
  cca_clusters == 1 ~ "MOL5/6", 
  cca_clusters == 2 ~ "MOL2", 
  cca_clusters == 3 ~ "MOL5/6",
  cca_clusters == 4 ~ "MOL5/6",
  cca_clusters == 5 ~ "MOL5/6",
  cca_clusters == 6 ~ "MOL5/6",
  cca_clusters == 7 ~ "MOL2", 
  cca_clusters == 8 ~ "MOL5/6", 
  cca_clusters == 9 ~ "MOL2", 
  cca_clusters == 10 ~ "MOL_DA1", 
  cca_clusters == 11 ~ "MOL5/6", 
  cca_clusters == 12 ~ "MOL_DA2", 
  cca_clusters == 13 ~ "NFOLs",
  cca_clusters == 14 ~ "MOL_IFN", 
  cca_clusters == 15 ~ "MOL5/6"
), levels = c("NFOLs", "MOL2", "MOL5/6","MOL_IFN",  "MOL_DA1", "MOL_DA2")))

oligo.plots[["UMAP_intOLSubtypes"]] <- merged.seurat %>% DimPlot(
  reduction = "umap.cca", group.by = "intOLSubtypes", cols = col.list$intOLSubtypes, pt.size = 1.2, shuffle = T
) & NoAxes()
```

save integrated annotated Zucha 2023 OLs
```{r save int anno seurat}
saveRDS(merged.seurat, file = "data/Seurat_OLs_integrated_annotated.Rds")
```

## Integrated scsn OLs
### sctransformed-centered analysis

```{r scsn seurat load}
oligoint.seurat <- readRDS("data/seurat_OLIGO_integrated_UMAP_named.rds")
oligoint.seurat %<>% SetIdent(value = "scsnOLAnno")
oligoint.seurat@active.ident %>% levels
```

Clustering of the integrated dataset
```{r clustering}
DefaultAssay(oligoint.seurat) <- "integrated"

oligoint.seurat %>% UMAPPlot(label = T)
oligoint.seurat %<>% FindClusters(resolution = 1.6, verbose = T, algorithm = 4) # min cluster size of 130 cells
oligoint.seurat %>% UMAPPlot(label = T)
```

Annotation of the annotated clusters
```{r cluster annotation}
DefaultAssay(oligoint.seurat) <- "SCT"
oligoint.seurat %<>% SetIdent(value = "integrated_snn_res.1.6")

oligoint.seurat[["scsnOLAnno"]] <- oligoint.seurat[["integrated_snn_res.1.6"]]
oligoint.seurat[["scsnOLAnno"]] <- case_when(
  oligoint.seurat$scsnOLAnno == "14" ~ "NFOL",
  
  oligoint.seurat$scsnOLAnno == "1" ~ "MOL2",
  oligoint.seurat$scsnOLAnno == "7" ~ "MOL2",
  oligoint.seurat$scsnOLAnno == "12" ~ "MOL2",
  
  oligoint.seurat$scsnOLAnno == "2" ~ "MOL5/6",
  oligoint.seurat$scsnOLAnno == "3" ~ "MOL5/6",
  oligoint.seurat$scsnOLAnno == "4" ~ "MOL5/6",
  oligoint.seurat$scsnOLAnno == "5" ~ "MOL5/6",
  oligoint.seurat$scsnOLAnno == "6" ~ "MOL5/6",
  oligoint.seurat$scsnOLAnno == "8" ~ "MOL5/6",
  oligoint.seurat$scsnOLAnno == "9" ~ "MOL5/6",
  oligoint.seurat$scsnOLAnno == "15" ~ "MOL5/6",
  oligoint.seurat$scsnOLAnno == "16" ~ "MOL5/6",
  
  oligoint.seurat$scsnOLAnno == "10" ~ "MOL_IFN",
  oligoint.seurat$scsnOLAnno == "11" ~ "MOL_DA1",
  oligoint.seurat$scsnOLAnno == "13" ~ "MOL_DA2",
  ) %>% factor(levels = c("NFOL", "MOL2", "MOL5/6", "MOL_IFN", "MOL_DA1", "MOL_DA2")); oligoint.seurat %<>% SetIdent(value = "scsnOLAnno")

oligoint.seurat$Condition %<>% factor(levels = c("sham", "D1", "D3", "D7"))

oligoint.seurat %>% UMAPPlot(cols = col.list[["ol_cols"]])

#saveRDS(oligoint.seurat, file = "data/seurat_OLIGO_integrated_UMAP_named.Rds")
```

UMAP clusters. Mito and ribo content
```{r scsn UMAP}
oligoint.seurat <- readRDS(file = "data/seurat_OLIGO_integrated_UMAP_named.Rds")

(oligoint.seurat %>% 
  UMAPPlot(group.by = "scsnOLAnno", pt.size = 1, label = F, label.box = T, cols = col.list[["ol_cols"]]) + 
  NoLegend() + 
  theme(plot.background = element_blank(), panel.background = element_blank())) %>% ggsave(filename = file.path(ws, "UMAPPlot_scsn_ol.tiff"), device = 'tiff', dpi = 300, units = 'cm', width = 10, height = 12)

## sc vs sn
(oligoint.seurat %>% 
  UMAPPlot(group.by = "orig.ident", pt.size = 1, label = F, label.box = T, cols = c("#525475", "#E6B188")) + 
  NoLegend() + 
  theme(plot.background = element_blank(), panel.background = element_blank())) %>% ggsave(filename = file.path(ws, "UMAPPlot_scsn_ol_origident.tiff"), device = 'tiff', dpi = 300, units = 'cm', width = 10, height = 12)

## condition
(oligoint.seurat %>% 
  UMAPPlot(group.by = "Condition", pt.size = 1, label = F, label.box = T, cols = c("sham" = "#525475", "D1" = "#F1D7AB", "D3" = "#E6B188", "D7" = "#D18080")) + 
  NoLegend() + 
  theme(plot.background = element_blank(), panel.background = element_blank())) %>% ggsave(filename = file.path(ws, "UMAPPlot_scsn_ol_conditions.tiff"), device = 'tiff', dpi = 300, units = 'cm', width = 10, height = 12)


DefaultAssay(oligoint.seurat) <- "SCT"
## percent ribo
(oligoint.seurat %>% FeaturePlot(features = "percent.rib", label = F, pt.size = 0.25, cols = brewer.pal(n = 9, name = "Blues")[3:9]) & 
    theme(text = element_text(family = "OpenSans"))) %>% 
    ggsave(filename = paste0(ws, "/FeaturePlot_scsnOL_percentRib.tiff"), device = 'tiff', units = "cm", dpi = 300, width = 8, height = 8) ## The legend corresponds to the color order, not absolute values. 
oligoint.seurat$percent.rib %>% summary


## percent mt
(oligoint.seurat %>% FeaturePlot(features = "percent.mt", label = F, pt.size = 0.25, cols = brewer.pal(n = 9, name = "Blues")[3:9]) & 
    theme(text = element_text(family = "OpenSans"))) %>% 
    ggsave(filename = paste0(ws, "/FeaturePlot_scsnOL_percentMt.tiff"), device = 'tiff', units = "cm", dpi = 300, width = 8, height = 8) ## The legend corresponds to the color order, not absolute values. 
oligoint.seurat$percent.mt %>% summary
```

Mito and ribo content continued.
```{r ribo mito}
oligoint.seurat[[]] %>% names ## percent.mt and percent.rib are calculated

# mitochondrial content
oligoint.seurat %>% FeaturePlot(
  features = "percent.mt",
  pt.size = 0.25,
  max.cutoff = 'q99',
  cols = brewer.pal(n = 9, name = "Blues")[3:9]
)

seurat.plots[["VlnPlot_intOLs_MT"]] <-
  oligoint.seurat %>%
  enh_vlnplot(
    feature = "percent.mt",
    grouping = "scsnOLAnno",
    colors = col.list[["intOLSubtypes"]],
    compare_means = F
  ) +
  theme_mk +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  remove_grid +
  NoLegend() +
  ylab("Mitochondrial RNA content (%)") +
  ggtitle(NULL)

print(seurat.plots[["VlnPlot_intOLs_MT"]])

seurat.plots[["VlnPlot_intOLs_MT"]] %>% 
  png_save_show(
    file = file.path(ws, "VlnPlot_intOLs_MT.png"), 
    show_plot = F, 
    width = 90, 
    height = 90
    )


# ribosomal content
oligoint.seurat %>% FeaturePlot(
  features = "percent.rib",
  pt.size = 0.25,
  max.cutoff = 'q99',
  cols = brewer.pal(n = 9, name = "Blues")[3:9]
) %>%
  ggsave(
    filename = file.path(ws, "FeaturePlot_scsn_oligo_percentRIB.tiff"),
    device = 'tiff',
    dpi = 300,
    width = 10,
    height = 10,
    units = 'cm'
  )

seurat.plots[["VlnPlot_intOLs_RIB"]] <-
  oligoint.seurat %>%
  enh_vlnplot(
    feature = "percent.rib",
    grouping = "scsnOLAnno",
    colors = col.list[["intOLSubtypes"]],
    compare_means = F
  ) +
  theme_mk +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  remove_grid +
  NoLegend() +
  ylab("Ribosomal RNA content (%)") +
  ggtitle(NULL)

print(seurat.plots[["VlnPlot_intOLs_RIB"]])

seurat.plots[["VlnPlot_intOLs_RIB"]] %>% 
  png_save_show(
    file = file.path(ws, "VlnPlot_intOLs_RIB.png"), 
    show_plot = F, 
    width = 90, 
    height = 90
    )
```

Cluster proportions - conditions
```{r scsn proportions}
## Cluster proportions
ids <- levels(oligoint.seurat@meta.data$Condition); categories <- levels(oligoint.seurat@meta.data$scsnOLAnno); counts <- matrix(nrow=length(ids), ncol=length(categories)); rownames(counts) <- ids; colnames(counts) <- categories
for (i in seq_along(ids)) {
  for (j in seq_along(categories)) {
    count <- oligoint.seurat@meta.data %>%
      filter(Condition == ids[i], scsnOLAnno == categories[j]) %>%
      nrow()

    counts[i, j] <- count
  }
}; rm(i,j)

counts_mm_0 <- reshape2::melt(counts, id = "rownames") ## total count proportions
ggplot(data = counts_mm_0, aes(x = Var2, y = value, fill = Var1)) + 
  geom_col(position = "dodge") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  ylab('Proportion of cells')
counts <- counts/rowSums(counts)
counts_mm_1 <- reshape2::melt(counts, id = "rownames")


p <-  ggplot(data = counts_mm_1, aes(x = Var2, y = value, fill = Var1)) + ## percent count proportions
    geom_col(position = "dodge") +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 105, hjust = 1, vjust = 0), axis.text.y = element_text(angle = 90, hjust = 0.5, vjust = 0.5)) +
    ylab('Proportion of cells') +
    xlab("") + ggtitle("") + NoLegend() +
    theme(axis.text = element_text(family = "OpenSans", size = 35, color = "black"),
        axis.title = element_text(family = "OpenSans", size = 35),
        plot.background = element_rect(fill = 'transparent', color = NA),
        panel.background = element_rect(fill = 'transparent', color = NA)) +
    scale_fill_manual(values = c("#525475", "#F1D7AB", "#E6B188", "#D18080")); p; 
p %>% ggsave(filename = file.path(ws, "BarPlot_PercentProportions_OurSCSNint_Oligo_bySample.tiff"), device = 'tiff', units = "cm", dpi = 300, width = 12, height = 7, bg = 'transparent')

## stacked barplots
library(dittoSeq)
dittoBarPlot(object = oligoint.seurat, 
                       group.by = "cell_type_3", 
                       scale = "percent",
                       var = 'orig.ident', color.panel = brewer.pal(n = 8, name = "Dark2")[c(3:4)], x.labels.rotate = T)


oligoint.seurat %>% dittoBarPlot(var = "cell_type_2", group.by = "cell_type_3", scale = "percent")
```

Markers, dotplot
```{r scsn markers, dotplot}
oligoint.seurat@active.ident %>% levels
oligoint.seurat %<>% SetIdent(value = "scsnOLAnno")

## Markers
oligoint.seurat %<>% PrepSCTFindMarkers()
oligoint.markers.all <- FindAllMarkers(oligoint.seurat, assay = 'SCT', logfc.threshold = 0.4, verbose = T, only.pos = T) %>% 
  arrange(desc(avg_log2FC)) %>% 
  filter(p_val_adj < 0.01) %>% 
  split(f = .[["cluster"]])
write.xlsx(oligoint.markers.all, file = file.path(ws, "Markers_intSCSN_OLs.xlsx"), overwrite = T) ## save the file



## Dotplot
oligoint.seurat %>% FeaturePlot(features = "Tnfrsf1a", pt.size = 1.5, label = T)
oligoint.seurat$scsnOLAnno %>% levels
feature <- 'Tnfrsf12a'; list('A' = oligoint.seurat %>% VlnPlot(features = feature),
                        'B' = oligoint.seurat %>% FeaturePlot(features = feature, pt.size = 1.2)) %>% gridExtra::grid.arrange(grobs = ., ncol = 2)

features_test <- c('Plp1', 'Mag', 'Mog', 'Olig1', 'Apod', 'Apoe', ## OL lineage
                   'Nckap5','Tcf7l2', 'Prom1',   # NFOL
                   'Sepp1', 'S100b', 'Hopx',  # MOL2
                   'Opalin', 'Ptgds',  # MOL5/6
                   'Cd9', 'C4b','Ifi27', 'H2-D1', 'B2m',  # MOL IFN
                   'Trf', 'Serpina3n', 'Klk6', # MOL DA1
                   'Cd63', 'Eif1', 'Cct5', 'Ctsd', 'Ddit3', 'Cdkn1a',  'Tnfrsf12a', "Tnfrsf1a"  # MOL DA2
                   )

p <-
  DotPlot(
    oligoint.seurat,
    assay = "RNA",
    features = features_test,
    group.by = "scsnOLAnno",
    cols = c("#FFF7FB", "#4D3D70")
  ) + 
  labs(x = NULL, y = NULL) +
  theme_mk +
  remove_grid + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1), 
        axis.ticks = element_blank())
print(p)

png_save_show(
  plot = p, 
  file = file.path(ws, "DotPlot_scsn_oligodendrocytes_markers.png"), 
  show_plot = TRUE, 
  dpi = 1000,
  height = 80, 
  width = 180)


## Individual gene featureplot
features <- c('Serpina3n', 'C4b', 'Klk6', 'Gadd45b', 'Tnfrsf1a', 'Eif4a1', 'Cebpd', 'Ifi27')
lapply(features, function(feature){
  (oligoint.seurat %>% FeaturePlot(features = feature, label = F, pt.size = 0.5, max.cutoff = 'q99') & 
    theme(text = element_text(family = "OpenSans")) & 
    viridis::scale_color_viridis(option = "rocket", direction = -1, alpha = 1, end = 0.99)) %>% 
    ggsave(filename = paste0(ws, "/FeaturePlot_scsnOL_gene_", feature, ".tiff"), device = 'tiff', units = "cm", dpi = 300, width = 8, height = 8)
})
```

Gene ontology analysis
```{r GO analysis}
library(clusterProfiler); library("org.Mm.eg.db", character.only = TRUE)
organism <- org.Mm.eg.db; background <- rownames(oligoint.seurat[["SCT"]])

oligoint.markers.all <- read_all_sheets(file = file.path(ws, "Markers_intSCSN_OLs.xlsx")) ## read the file

## GO analysis
oligoint.seurat@active.ident %>% table
pops <- c("MOL_IFN", "MOL_DA1", "MOL_DA2")
diff.reac.list <- lapply(pops, function(pops){oligoint.markers.all[[pops]][["gene"]] %>% .[grep("^Rp|^mt", x = ., ignore.case = F, invert = T)] ## removing ribosomal and mito genes 
  }); names(diff.reac.list) <- pops


ORA.list <- lapply(diff.reac.list, function(x) enrichGO(gene          = x , # list of genes 
                                                      universe      = background, # 'potential DEGs'
                                                      keyType       = "SYMBOL", # or ENSEMBL
                                                      OrgDb         = organism,
                                                      ont           = "ALL", # BP = biological process, MF = molecular function, CC = cellular component
                                                      pAdjustMethod = "fdr",
                                                      pvalueCutoff  = 0.01, #
                                                      qvalueCutoff  = 0.2,
                                                      readable      = F,
                                                      minGSSize     = 3) %>% clusterProfiler::simplify()
                   )

ORA.list %>% write.xlsx(file = file.path(ws, "ORAlist_scsn_MOL_DA.xlsx"), overwrite = T)


## selected representative GO terms
ORA.list <- read_all_sheets(file.path(ws, "ORAlist_scsn_MOL_DA.xlsx"))
terms.list <- list(
  "MOL_IFN" = c(
    "GO:0006897", # endocytosis (BP)
    "GO:0006959", # humoral response (BP)
    "GO:0010039", # response to iron ion (BP)
    "GO:0042612" # MHC class I (CC)
  ),
  "MOL_DA1" = c(
    "GO:0043209", # myelin sheath (CC)
    "GO:0003779", # actin binding (MF)
    "GO:0005200" # structural component of cytoskeleton (MF)
  ),
  "MOL_DA2" = c(
    "GO:0006413", # translation initiation (BP)
    "GO:0006986", # response to unfolded protein (BP)
    "GO:0097193", # intrinsic apoptotic pathway (BP)
    "GO:0006091", # generation of precursor metabolites and energy (BP)
    "GO:1901214", # regulation of neuron death (BP)
    "GO:0052547" # regulation of peptidase pathway (BP)
  )
)

GO.plot <- lapply(names(terms.list) %>% rev, function(population){
  subset.ORA <- ORA.list[[population]] %>% filter(ID %in% terms.list[[population]])
  ORA.data <- data.frame(
    'Description' = subset.ORA[["Description"]] %>% stringr::str_to_title() %>% stringr::str_replace_all(c("Mhc" = "MHC", "Atp" = "ATP")),
    'Pval' = subset.ORA[["p.adjust"]] %>% log10 %>% '*'(-1),
    'Population' = rep(paste0(population), nrow(subset.ORA))) %>% arrange(Pval)
  }
); GO.plot %<>% dplyr::bind_rows(); GO.plot$Description %<>% factor(levels = unique(.))


## barplot
(ggplot(data = GO.plot, aes(x = Description, y = Pval, fill = Population)) + 
    geom_bar(stat = "identity", width = 0.6, size = 0.2, colour = "black", alpha = 0.75) +
    scale_fill_manual(values = col.list[["ol_cols"]]) +
    ylim(0, max(GO.plot$Pval) + 1) + 
    geom_hline(yintercept = 1.3, size = 0.2) +
    coord_flip() +
    theme(text = element_text(family = "OpenSans", size = 40), 
          axis.text.x = element_text(vjust = 1, hjust = 1,size = 40),
          axis.text.y = element_text(vjust = 0.5, hjust = 1, size = 40, color = 'black'),
          axis.ticks = element_blank(),
          panel.background = element_blank(), #transparent panel bg
          plot.background = element_blank(), #transparent plot bg 
          panel.grid.major.x = element_blank(), 
          panel.grid.major.y = element_line(colour = "#A1A19C", size = 0.2)
          ) + NoLegend() +
          xlab("") + ylab(expression(paste("-log(Padj)")))) %>% ggsave(filename = paste0(ws, "/Barplot_scsn_OL_GO.tiff"), device = "tiff", dpi = 300, width = 8.5, height = 6)
```

ScSn signatures in the spatial data.
```{r scsn profiles in spatial}
markers_gos <- read_all_sheets(file = file.path("data", "SupplementaryTable6_sn_sc_Markers_GOanalysis.xlsx")) ## read the file
names(markers_gos)
pops <- c("intSNSC_MOL_IFN", "intSNSC_MOL_DA1", "intSNSC_MOL_DA2")
diff.reac.list <- lapply(pops, function(pops){markers_gos[[pops]][["gene"]] %>% .[grep("^Rp|^mt", x = ., ignore.case = F, invert = T)] ## removing ribosomal and mito genes 
  }); names(diff.reac.list) <- pops

reac.list <- list(
  "MOL_IFN" = diff.reac.list[["intSNSC_MOL_IFN"]] %>% head(20),
  "MOL_DA1" = diff.reac.list[["intSNSC_MOL_DA1"]] %>% head(20),
  "MOL_DA2" = diff.reac.list[["intSNSC_MOL_DA2"]] %>% head(20)
)
spatial.seurat %<>% AddModuleScore(features = reac.list, name = names(reac.list), assay = "SCT")
spatial.seurat@meta.data %>% names ## modules scores were added (the last three)
features <- spatial.seurat@meta.data %>% names %>% tail(3)

## Spatial plots
oligo.plots[["SpatialPlot_reactiveOLs"]] <- 
spatial.seurat %>% 
  SpatialPlot(
    features = features,
    images = sections,
    crop = TRUE, 
    image.alpha = 0, 
    ncol = length(sections), 
    pt.size.factor = 2.5, 
    stroke = 0
  ) & 
  theme_mk &
  remove_grid &
  NoAxes() & 
  theme(
    legend.position = "none", 
    panel.border = element_blank()) & 
  scale_fill_gradientn(colours = c("grey50", "#FFFFFF", "#211E61"))

print(oligo.plots[["SpatialPlot_reactiveOLs"]])

png_save_show(
  plot = oligo.plots[["SpatialPlot_reactiveOLs"]], 
  file = file.path(ws, "SpatialPlot_reactiveOLs.png"), 
  show_plot = FALSE, 
  dpi = 1000, 
  width = 120, 
  height = 80
)
```

```{r Metanalysis}
oligoint.seurat %<>% SetIdent(value = "scsnOLAnno")
oligoint.seurat %>% UMAPPlot(label = T, pt.size = 2, label.box = T)
oligoint.seurat %>% UMAPPlot(label = T, group.by = 'cell_type_3')

## markers
oligoint.markers.all <- read_all_sheets(file = file.path(ws, "Markers_intSCSN_OLs.xlsx")) ## read in the markers
oligoint.seurat@active.ident %>% table

oligoint.seurat %>% UMAPPlot(label = T, split.by = 'orig.ident', pt.size = 2)
oligoint.seurat %>% UMAPPlot(label = T, split.by = 'orig.ident', group.by = 'Condition', pt.size = 2)
oligoint.seurat %>% UMAPPlot(label = T, split.by = 'Condition', pt.size = 2)


### MOL5/6
features <- c('Ptgds', 'Il33')
FeaturePlot(oligoint.seurat, features = features, label = T, pt.size = 2, min.cutoff = 'q5')
oligoint.seurat %>% VlnPlot(features = features)

### MOL2
features <- c('Klk6', 'Hopx')
FeaturePlot(oligoint.seurat, features = features, label = T, pt.size = 2, min.cutoff = 'q5')

### MOL_DA1
features <- c('C4b', 'Serpina3n', 'Tagln2')
FeaturePlot(oligoint.seurat, features = features, label = T, pt.size = 2)
oligoint.seurat %>% VlnPlot(features = features)

### MOL_DA2
features <- c('Cdkn1a', 'Tnfrsf12a', 'Bax')
FeaturePlot(oligoint.seurat, features = features, label = T, pt.size = 2)
oligoint.seurat %>% VlnPlot(features = features)

### MOL_IFN
features <- c('H2-D1', 'Ifit1', 'Stat1', 'Bst2')
FeaturePlot(oligoint.seurat, features = features, label = T, pt.size = 2)
oligoint.seurat %>% VlnPlot(features = features)


## select representative dis assoc oligo datasets capturing a spectrum of pathologies. 
metadata.list <- read_all_sheets(file = "data/SupplementaryTable4_Metanalysis_GlialReactivityMarkers.xlsx") %>% 
  .[names(.) %>% grep(pattern = "MOL|_OL_|DOL", ignore.case = T)] %>%  ## keep the OL-related populations
  lapply(., function(df) df %>% pull()) ## change the dataframes into vectors

oligoint.seurat %<>% AddModuleScore(features = metadata.list, name = names(metadata.list))
oligoint.seurat[[]] %>% names
features <- oligo.seurat[[]] %>% names %>% tail(length(metadata.list))

##feature plots
lapply(features, function(feature){
  feature.name <- feature %>% stringr::str_sub(end = -2)
  (oligoint.seurat %>% FeaturePlot(features = feature, label = F, pt.size = 0.25, max.cutoff = 'q99') & 
    theme(text = element_text(family = "OpenSans")) & 
    viridis::scale_color_viridis(option = "rocket", direction = -1, alpha = 1, end = 0.99)
    ) %>% 
    ggsave(filename = paste0(ws, "/FeaturePlot_scsnOL_metanalysis_", feature.name, ".tiff"), device = 'tiff', units = "cm", dpi = 300, width = 8, height = 8)
})


##vln plots
lapply(features, function(feature){
  feature.name <- feature %>% stringr::str_sub(end = -2) %>% stringr::str_replace_all(pattern = "_", replacement = " ")
  print(feature.name)
  (oligoint.seurat %>% 
      VlnPlot(features = feature, cols = col.list[["ol_cols"]] , pt.size = 0) & 
      ylab('Expression signature') &
      xlab("") &
      labs(subtitle = feature.name, title = NULL) &
      theme(text = element_text(family = "OpenSans"),
            title = element_text(face = 'bold', size = 20),
            axis.title = element_text(size = 20),
            axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 20), 
            axis.text.y = element_text(hjust = 0.5, vjust = 0.5, size = 20)) &
      NoLegend()      
  ) %>% ggsave(filename = paste0(ws, "/VlnPlot_Metanalysis_ourSCSN_", feature.name, ".tiff"), device = 'tiff', dpi = 300, units = 'cm', width = 6, height = 5)
})
  oligoint.seurat %>% VlnPlot(features = features)


#### metanalysis statistical signif
features
OLsplitlist <- oligoint.seurat@meta.data %>% group_split(scsnOLAnno); names(OLsplitlist) <- oligoint.seurat$scsnOLAnno %>% levels

wilcox.meta <- lapply(features, function(feature){
  temp <- lapply(c('MOL_IFN','MOL_DA1', 'MOL_DA2'), function(OL){
    test.res <- wilcox.test(x = OLsplitlist[[OL]][[feature]],
                            y = c(OLsplitlist[['MOL2']][[feature]], OLsplitlist[['MOL5/6']][[feature]]), conf.int = T)
    test.pval <- test.res[["p.value"]]
  }); names(temp) <- c('MOL_IFN','MOL_DA1', 'MOL_DA2')
  return(temp)
}); names(wilcox.meta) <- features; wilcox.meta %<>% unlist %>% p.adjust(method = 'fdr'); wilcox.meta %>% .[. < 0.01] 

#### in spatial
## plot the spatial plots separately
spatial.seurat %<>% AddModuleScore(features = metadata.list, name = names(metadata.list))
lapply(features, function(feature){
  feature.name <- feature %>% stringr::str_sub(end = -2)
  print(feature.name)
  (spatial.seurat %>% 
      SpatialPlot(features = feature, crop = F, pt.size = 1.35, alpha = c(0.4, 1), images = c('sham', 'D1', 'D3', 'D7'), max.cutoff = 'q99', image.alpha = 0.4, stroke = 0) & 
      theme(text = element_text(family = plot.font, size = 9), legend.key.size = unit(0.2, 'cm')) & NoLegend() &
      ggplot2::scale_fill_gradientn(colours = RColorBrewer::brewer.pal(name = "RdBu", n = 100) %>% rev, limits = c(min(spatial.seurat[[feature]]), max(spatial.seurat[[feature]])))) %>% 
  ggsave(filename = paste0(ws, "/SpatialPlot_Metanalysis_", feature.name, ".tiff"), device = 'tiff', dpi = 300, width = 18, height = 6, units = 'cm')
})

```

Immunohistochemistry validation
```{r immuno validation}
# Serpina3n
immuno <- 
  read_all_sheets(
    file = file.path("results", "PNAS", "Tables", "MCAO_IHC_Validations.xlsx")
    ) %>% 
  .[["Plp1_Serpina3n"]] %>% 
  mutate(Proportion = Proportion * 100) %>% 
  dplyr::mutate(Condition = factor(Condition, levels = sections)) %>% 
  group_by(SampleName, Condition) %>% 
  dplyr::summarise(
    mean_colocalization = mean(Proportion), 
    .groups = "drop"
  )

p.values <- 
  pairwise.wilcox.test(
    x = immuno$mean_colocalization, 
    g = immuno$Condition, 
    p.adjust.method = "BH", 
    paired = F)

immuno_summary <- 
  immuno %>% 
  group_by(Condition) %>% 
  dplyr::summarise(
    sd_colocalization = stats::sd(mean_colocalization),
    mean_colocalization = mean(mean_colocalization)
  )

# stat test control vs injured
stat_test <- 
  immuno %>% 
  mutate(Injured =  if_else(Condition == "Ctrl", "NoMCAO", "MCAO")) %>% 
  wilcox.test(mean_colocalization ~ Injured, data = .)
print(stat_test)


# plot
seurat.plots[["Barplot_Immuno_OL_Serpina"]] <- 
  ggplot(
    data = immuno, 
    aes(y = mean_colocalization, x = Condition)
    ) + 
  geom_bar(data = immuno_summary, stat = "identity", width = 0.6, colour = "black", alpha = 0.75, fill = "#4D3D70", size = 0.5) +
  geom_errorbar(
    data =  immuno_summary, 
    aes(ymin = mean_colocalization, 
        ymax = mean_colocalization + sd_colocalization), width = 0.1, color = "black", size = 0.5) +
  geom_jitter(position = position_jitter(0.2), color = 'black', size= 1.5, shape = 1, stroke = 0.5) + 
  annotate(geom = "text", x = 1, y = 1, label = paste0("P = ", stat_test$p.value %>% round(3)), vjust = -0.5, hjust = 0, size = 2.45) +
  ylim(-0.1, 1.5) +
  xlab(NULL) + 
  ylab("Plp1 + SERPINA3N co-localization (%)") +
  theme_mk + 
  remove_grid + 
  theme(panel.grid.major.y = element_line(linewidth = 0.1, colour = "grey"))
print(seurat.plots[["Barplot_Immuno_OL_Serpina"]])

png_save_show(
  plot = seurat.plots[["Barplot_Immuno_OL_Serpina"]], 
  file = file.path(ws, "Barplot_Immuno_OL_Serpina.png"), 
  show_plot = FALSE, 
  dpi = 1000, 
  height = 70, 
  width = 50
)

# B2m
immuno <- 
  read_all_sheets(
    file = file.path("results", "PNAS", "Tables", "MCAO_IHC_Validations.xlsx")
    ) %>% 
  .[["Plp1_B2m"]] %>% 
  mutate(Proportion = Proportion * 100) %>% 
  dplyr::mutate(Condition = factor(Condition, levels = sections)) %>% 
  group_by(SampleName, Condition) %>% 
  dplyr::summarise(
    mean_colocalization = mean(Proportion), 
    .groups = "drop"
  )

immuno_summary <- 
  immuno %>% 
  group_by(Condition) %>% 
  dplyr::summarise(
    sd_colocalization = stats::sd(mean_colocalization),
    mean_colocalization = mean(mean_colocalization)
  )

# stat test
## pairwise
p.values <- 
  pairwise.wilcox.test(
    x = immuno$mean_colocalization, 
    g = immuno$Condition, 
    p.adjust.method = "BH", 
    paired = F)
p.values

# control vs injured
stat_test <- 
  immuno %>% 
  mutate(Injured =  if_else(Condition == "Ctrl", "NoMCAO", "MCAO")) %>% 
  wilcox.test(mean_colocalization ~ Injured, data = .)
print(stat_test)

# plot
seurat.plots[["Barplot_Immuno_OL_B2m"]] <- 
  ggplot(
    data = immuno, 
    aes(y = mean_colocalization, x = Condition)
    ) + 
  geom_bar(data = immuno_summary, stat = "identity", width = 0.6, colour = "black", alpha = 0.75, fill = "#A23C8A", size = 0.5) +
  geom_errorbar(
    data =  immuno_summary, 
    aes(ymin = mean_colocalization, 
        ymax = mean_colocalization + sd_colocalization), width = 0.1, color = "black", size = 0.5) +
  geom_jitter(position = position_jitter(0.2), color = 'black', size= 1.5, shape = 1, stroke = 0.5) + 
  annotate(geom = "text", x = 1, y = 6, label = paste0("P = ", stat_test$p.value %>% round(3)), vjust = -0.5, hjust = 0, size = 2.45) +
  ylim(-0.1, 11) +
  xlab(NULL) + 
  ylab("Plp1 + B2M co-localization (%)") +
  theme_mk + 
  remove_grid + 
  theme(panel.grid.major.y = element_line(linewidth = 0.1, colour = "grey"))
print(seurat.plots[["Barplot_Immuno_OL_B2m"]])

png_save_show(
  plot = seurat.plots[["Barplot_Immuno_OL_B2m"]], 
  file = file.path(ws, "Barplot_Immuno_OL_B2m.png"), 
  show_plot = FALSE, 
  dpi = 1000, 
  height = 70, 
  width = 50
)

# Tnfrsf1a
immuno <- 
  read_all_sheets(
    file = file.path("results", "PNAS", "Tables", "MCAO_IHC_Validations.xlsx")
    ) %>%
  .[["Plp1_Tnfrsf1a"]] %>% 
  mutate(Proportion = Proportion * 100) %>% 
  dplyr::mutate(Condition = factor(Condition, levels = sections)) %>% 
  group_by(SampleName, Condition) %>% 
  dplyr::summarise(
    mean_colocalization = mean(Proportion), 
    .groups = "drop"
  )

immuno_summary <- 
  immuno %>% 
  group_by(Condition) %>% 
  dplyr::summarise(
    sd_colocalization = stats::sd(mean_colocalization),
    mean_colocalization = mean(mean_colocalization)
  )

# stat test
## pairwise
p.values <- 
  pairwise.wilcox.test(
    x = immuno$mean_colocalization, 
    g = immuno$Condition, 
    p.adjust.method = "BH", 
    paired = F)
p.values

## control vs injured
stat_test <- 
  immuno %>% 
  mutate(Injured =  if_else(Condition == "Ctrl", "NoMCAO", "MCAO")) %>% 
  wilcox.test(mean_colocalization ~ Injured, data = .)
print(stat_test)

# plot
seurat.plots[["Barplot_Immuno_OL_Tnfrsf1a"]] <- 
  ggplot(
    data = immuno, 
    aes(y = mean_colocalization, x = Condition)
    ) + 
  geom_bar(data = immuno_summary, stat = "identity", width = 0.6, colour = "black", alpha = 0.75, fill = "#211E61", size = 0.5) +
  geom_errorbar(
    data =  immuno_summary, 
    aes(ymin = mean_colocalization, 
        ymax = mean_colocalization + sd_colocalization), width = 0.1, color = "black", size = 0.5) +
  geom_jitter(position = position_jitter(0.2), color = 'black', size= 1.5, shape = 1, stroke = 0.5) + 
  annotate(geom = "text", x = 1, y = 50, label = paste0("P = ", stat_test$p.value %>% round(3)), vjust = -0.5, hjust = 0, size = 2.45) +
  ylim(-1, 90) +
  xlab(NULL) + 
  ylab("Plp1 + TNFRSF1A co-localization (%)") +
  theme_mk + 
  remove_grid + 
  theme(panel.grid.major.y = element_line(linewidth = 0.1, colour = "grey"))
print(seurat.plots[["Barplot_Immuno_OL_Tnfrsf1a"]])

png_save_show(
  plot = seurat.plots[["Barplot_Immuno_OL_Tnfrsf1a"]], 
  file = file.path(ws, "Barplot_Immuno_OL_Tnfrsf1a.png"), 
  show_plot = FALSE, 
  dpi = 1000, 
  height = 70, 
  width = 50
)

```





== Supplementary Figures ==

Full single-nucleus dataset
Summarizing plots of finalized sn dataset
```{r summary plots of whole sn dataset}
DefaultAssay(sn.seurat) <- "RNA" 
sn.seurat %<>% SetIdent(value = "Celltypes")

# UMAP of Celltypes
seurat.plots[["UMAP of Celltypes"]] <- sn.seurat %>% 
  DimPlot(reduction = "umap", pt.size = 1, label = T, group.by = 'Celltypes', cols = col.list$Celltypes) + NoLegend() + NoAxes()

# UMAP of Cell subtypes
seurat.plots[["UMAP of Cell Subtypes"]] <- sn.seurat %>% 
  DimPlot(reduction = "umap", pt.size = 1, label = F, group.by = 'CelltypesDetailed', cols = col.list$CelltypesDetailed) + NoLegend() + NoAxes()

# Cell type frequencies
seurat.plots[["Barplot of Celltypes"]] <- sn.seurat %>% 
  total_barplot(groupvar = "Celltypes", ylim = 2550)

# Cell Subtype frequencies
seurat.plots[["Barplot of CellSubtypes"]] <- sn.seurat %>% 
  total_barplot(groupvar = "CelltypesDetailed", ylim = 900)

# Cell type  proportions

# Markers dotplot
features <- c('Stmn1', 'Sox4', 'Ccnd2', 'Dcx',   # Neuroblast
                   'Usp29', 'Snhg11', 'Nhs', # Other Neurons
                   'Gad1', 'Gad2', # GABAergic
                   'Meg3', 'Rbfox3',  'Nrgn', 'Slc17a6', 'Slc17a7',  # Glutamatergic
                   'Ccdc153', 'Foxj1',  # Ependymal
                   'Aqp4',  'Aldh1l1', 'Slc1a3', 'Gfap',   # Astrocyte
                   'Pdgfra', 'Cspg4', 'Vcan', # OPC
                   'Mbp',  'Mobp', 'Plp1', 'Cldn11',    # Oligo
                   'Cx3cr1', 'Csf1r', # Microglia
                   'Slc47a1',  'Mgp', 'Ptgds',    # VLMC
                  'Pdgfrb',  'Flt1', 'Pecam1'    # Vasculature
                   )

seurat.plots[["Dotplot_Markers"]] <- sn.seurat %>% 
  dittoSeq::dittoDotPlot(vars = features, group.by = "Celltypes", scale = T)

# Displaying a specific gene
sn.seurat %>% 
  dittoSeq::dittoPlot(var = "Apoe", group.by = "Celltypes", plots = c("ridgeplot", "jitter") , assay = "RNA", slot = "data", color.panel = col.list$Celltypes) + theme(legend.position = "none")
```

Population proportions 
```{r proportions}
## proportions
library(dittoSeq)
sn.seurat$Condition %>% table

sn.seurat %>% dittoBarPlot(var = "Condition", group.by = "Celltypes", scale = "count")
sn.seurat %>% dittoBarPlot(var = "Condition", group.by = "Celltypes", scale = "percent")

sn.seurat[['Condition2']] <- case_when( ## simplifying the conditions
  sn.seurat$Condition == 'sham' ~ 'Ctrl',
               sn.seurat$Condition == 'D1' ~ 'D1',
               sn.seurat$Condition == 'D3' ~ 'D3',
               sn.seurat$Condition == 'D7' ~ 'D7') %>% factor(levels = c('Ctrl', 'D1', 'D3', 'D7'))

ids <- sn.seurat@meta.data$Condition2 %>% levels; categories <- sn.seurat@meta.data$cell_type_1_simplified %>% levels; counts <- matrix(nrow=length(ids), ncol=length(categories)); rownames(counts) <- ids; colnames(counts) <- categories
for (i in seq_along(ids)) {
  for (j in seq_along(categories)) {
    count <- sn.seurat@meta.data %>%
      filter(Condition2 == ids[i], cell_type_1_simplified == categories[j]) %>%
      nrow()

    counts[i, j] <- count
  }
}; rm(i,j)

counts_mm_0 <- reshape2::melt(counts, id = "rownames") ## total count proportions

ggplot(data = counts_mm_0, aes(x = Var2, y = value, fill = Var1)) + 
  geom_col(position = "dodge") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  ylab('Proportion of cells')
counts <- counts/rowSums(counts)
counts_mm_1 <- reshape2::melt(counts, id = "rownames") ## percent count proportions

## BarPlot on percentage of count proportions
seurat.plots[["BarPlot_sn_CellTypes1simp_bySample"]] <-  ggplot(data = counts_mm_1, aes(x = Var2, y = value, fill = Var1)) + ## percent count proportions
    geom_col(position = "dodge") +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 105, hjust = 1, vjust = 0), axis.text.y = element_text(angle = 90, hjust = 0.5, vjust = 0.5)) +
    ylab('Proportion of nuclei') +
    xlab("") + ggtitle("") + NoLegend() +
    theme(axis.text = element_text(family = "OpenSans", size = 25, color = "black"),
        axis.title = element_text(family = "OpenSans", size = 25),
        plot.background = element_rect(fill = 'transparent', color = NA),
        panel.background = element_rect(fill = 'transparent', color = NA)) +
    scale_fill_manual(values = c("#525475", "#F1D7AB", "#E6B188", "#D18080")); p


seurat.plots[["BarPlot_sn_CellTypes1simp_bySample"]] 
```


